<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Shaojun Xie" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Bioi611 prep monocle env - Lab note for UMD BIOI611</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Bioi611 prep monocle env";
        var mkdocs_page_input_path = "bioi611_prep_monocle_env.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/yaml.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/rust.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> Lab note for UMD BIOI611
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption"><span class="caption-text">Get ready for the course</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../basic_linux/">Basic Linux</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Bulk RNA-seq</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../bulkRNAseq_lab/">Read QC and alignment</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Phred_FQ/">Phred score in FASTQ</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_introR/">Minimal R Introduction</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_DESeq2_analysis/">DEG analysis using DESeq2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_DAVID/">Pathway analysis using DAVID</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_ballgown_example/">DE analysis at isoform-leve using ballgown (example data)</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_bulkRNA_SE_ballgown/">DE analysis at isoform-leve using ballgown (C. ele data)</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_long_read_transcriptome/">Long read transcriptome</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">scRNA-seq</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_analysis_10x_dataset_PBMC/">Initial analysis of 10x scRNA-seq data for human PBMC using cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_scRNA/">Downstream analysis of 10x scRNA-seq data for human PBMC using Seurat</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_scRNA_seq_cele_cellranger/">Initial analysis of 10x scRNA-seq data for C. elegans using cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_scRNA_cele/">Downstream analysis of 10x scRNA-seq data for C. elegans data</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bioi611_monocle_cele/">Downstream analysis - trajectory analysis using Monocle3</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">Lab note for UMD BIOI611</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Bioi611 prep monocle env</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/Bix4UMD/BIOI611_lab.git/edit/master/docs/bioi611_prep_monocle_env.md">Edit on Bix4UMD/BIOI611_lab</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <p><a href="https://colab.research.google.com/github/Bix4UMD/BIOI611_lab/blob/main/docs/bioi611_prep_monocle_env.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<pre><code class="language-R">if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))
install.packages(&quot;BiocManager&quot;)
BiocManager::install(version = &quot;3.20&quot;)
</code></pre>
<pre><code>Installing package into â€˜/usr/local/lib/R/site-libraryâ€™
(as â€˜libâ€™ is unspecified)

'getOption("repos")' replaces Bioconductor standard repositories, see
'help("repositories", package = "BiocManager")' for details.
Replacement repositories:
    CRAN: https://cran.rstudio.com

Bioconductor version 3.20 (BiocManager 1.30.25), R 4.4.2 (2024-10-31)

Installing package(s) 'BiocVersion'
</code></pre>
<pre><code class="language-R">## required by scater package
system(&quot;apt-get install libx11-dev libcairo2-dev&quot;) #, intern = TRUE)
</code></pre>
<pre><code class="language-R">BiocManager::install(c('BiocGenerics', 'DelayedArray', 'DelayedMatrixStats',
                       'limma', 'lme4', 'S4Vectors', 'SingleCellExperiment',
                       'SummarizedExperiment', 'batchelor', 'HDF5Array',
                       'terra', 'ggrastr'))
</code></pre>
<pre><code>'getOption("repos")' replaces Bioconductor standard repositories, see
'help("repositories", package = "BiocManager")' for details.
Replacement repositories:
    CRAN: https://cran.rstudio.com

Bioconductor version 3.20 (BiocManager 1.30.25), R 4.4.2 (2024-10-31)

Installing package(s) 'BiocGenerics', 'DelayedArray', 'DelayedMatrixStats',
  'limma', 'lme4', 'S4Vectors', 'SingleCellExperiment', 'SummarizedExperiment',
  'batchelor', 'HDF5Array', 'terra', 'ggrastr'

also installing the dependencies â€˜formatRâ€™, â€˜zlibbiocâ€™, â€˜lambda.râ€™, â€˜futile.optionsâ€™, â€˜matrixStatsâ€™, â€˜abindâ€™, â€˜XVectorâ€™, â€˜UCSC.utilsâ€™, â€˜GenomeInfoDbDataâ€™, â€˜assortheadâ€™, â€˜irlbaâ€™, â€˜rsvdâ€™, â€˜futile.loggerâ€™, â€˜snowâ€™, â€˜BHâ€™, â€˜beeswarmâ€™, â€˜viporâ€™, â€˜MatrixGenericsâ€™, â€˜IRangesâ€™, â€˜S4Arraysâ€™, â€˜SparseArrayâ€™, â€˜sparseMatrixStatsâ€™, â€˜statmodâ€™, â€˜minqaâ€™, â€˜nloptrâ€™, â€˜RcppEigenâ€™, â€˜GenomicRangesâ€™, â€˜Biobaseâ€™, â€˜GenomeInfoDbâ€™, â€˜igraphâ€™, â€˜BiocNeighborsâ€™, â€˜BiocSingularâ€™, â€˜BiocParallelâ€™, â€˜scuttleâ€™, â€˜ResidualMatrixâ€™, â€˜ScaledMatrixâ€™, â€˜beachmatâ€™, â€˜rhdf5â€™, â€˜rhdf5filtersâ€™, â€˜Rhdf5libâ€™, â€˜Cairoâ€™, â€˜ggbeeswarmâ€™, â€˜pngâ€™
</code></pre>
<pre><code class="language-R">install.packages(&quot;devtools&quot;)
devtools::install_github('cole-trapnell-lab/monocle3')
</code></pre>
<pre><code>Installing package into â€˜/usr/local/lib/R/site-libraryâ€™
(as â€˜libâ€™ is unspecified)

Downloading GitHub repo cole-trapnell-lab/monocle3@HEAD



sitmo        (NA -&gt; 2.0.2    ) [CRAN]
bitops       (NA -&gt; 1.0-9    ) [CRAN]
RCurl        (NA -&gt; 1.98-1.16) [CRAN]
proxy        (NA -&gt; 0.4-27   ) [CRAN]
wk           (NA -&gt; 0.9.4    ) [CRAN]
e1071        (NA -&gt; 1.7-16   ) [CRAN]
units        (NA -&gt; 0.8-5    ) [CRAN]
s2           (NA -&gt; 1.1.7    ) [CRAN]
classInt     (NA -&gt; 0.4-10   ) [CRAN]
sp           (NA -&gt; 2.1-4    ) [CRAN]
warp         (NA -&gt; 0.2.1    ) [CRAN]
parallelly   (NA -&gt; 1.39.0   ) [CRAN]
listenv      (NA -&gt; 0.9.1    ) [CRAN]
globals      (NA -&gt; 0.16.3   ) [CRAN]
future       (NA -&gt; 1.34.0   ) [CRAN]
lazyeval     (NA -&gt; 0.2.2    ) [CRAN]
RcppHNSW     (NA -&gt; 0.6.0    ) [CRAN]
gridExtra    (NA -&gt; 2.3      ) [CRAN]
RcppProgress (NA -&gt; 0.4.2    ) [CRAN]
dqrng        (NA -&gt; 0.4.1    ) [CRAN]
RSpectra     (NA -&gt; 0.16-2   ) [CRAN]
RcppAnnoy    (NA -&gt; 0.0.22   ) [CRAN]
FNN          (NA -&gt; 1.1.4.1  ) [CRAN]
biglm        (NA -&gt; 0.9-3    ) [CRAN]
deldir       (NA -&gt; 2.0-4    ) [CRAN]
sf           (NA -&gt; 1.0-19   ) [CRAN]
spData       (NA -&gt; 2.3.3    ) [CRAN]
slider       (NA -&gt; 0.3.2    ) [CRAN]
furrr        (NA -&gt; 0.3.1    ) [CRAN]
plyr         (NA -&gt; 1.8.9    ) [CRAN]
crosstalk    (NA -&gt; 1.2.1    ) [CRAN]
zoo          (NA -&gt; 1.8-12   ) [CRAN]
viridis      (NA -&gt; 0.6.5    ) [CRAN]
uwot         (NA -&gt; 0.2.2    ) [CRAN]
speedglm     (NA -&gt; 0.3-5    ) [CRAN]
spdep        (NA -&gt; 1.3-6    ) [CRAN]
slam         (NA -&gt; 0.1-55   ) [CRAN]
Rtsne        (NA -&gt; 0.17     ) [CRAN]
RhpcBLASctl  (NA -&gt; 0.23-42  ) [CRAN]
rsample      (NA -&gt; 1.2.1    ) [CRAN]
reshape2     (NA -&gt; 1.4.4    ) [CRAN]
RANN         (NA -&gt; 2.6.2    ) [CRAN]
pscl         (NA -&gt; 1.5.9    ) [CRAN]
plotly       (NA -&gt; 4.10.4   ) [CRAN]
pheatmap     (NA -&gt; 1.0.12   ) [CRAN]
pbmcapply    (NA -&gt; 1.5.1    ) [CRAN]
pbapply      (NA -&gt; 1.7-2    ) [CRAN]
lmtest       (NA -&gt; 0.9-40   ) [CRAN]
leidenbase   (NA -&gt; 0.1.31   ) [CRAN]
grr          (NA -&gt; 0.9.5    ) [CRAN]
ggrepel      (NA -&gt; 0.9.6    ) [CRAN]
assertthat   (NA -&gt; 0.2.1    ) [CRAN]


Skipping 30 packages ahead of CRAN: zlibbioc, XVector, SparseArray, S4Arrays, IRanges, S4Vectors, MatrixGenerics, BiocGenerics, GenomeInfoDbData, GenomeInfoDb, Rhdf5lib, rhdf5filters, DelayedArray, Biobase, sparseMatrixStats, beachmat, DelayedMatrixStats, SummarizedExperiment, GenomicRanges, BiocParallel, SingleCellExperiment, ScaledMatrix, rhdf5, ResidualMatrix, scuttle, BiocSingular, BiocNeighbors, limma, HDF5Array, batchelor

Installing 52 packages: sitmo, bitops, RCurl, proxy, wk, e1071, units, s2, classInt, sp, warp, parallelly, listenv, globals, future, lazyeval, RcppHNSW, gridExtra, RcppProgress, dqrng, RSpectra, RcppAnnoy, FNN, biglm, deldir, sf, spData, slider, furrr, plyr, crosstalk, zoo, viridis, uwot, speedglm, spdep, slam, Rtsne, RhpcBLASctl, rsample, reshape2, RANN, pscl, plotly, pheatmap, pbmcapply, pbapply, lmtest, leidenbase, grr, ggrepel, assertthat

Installing packages into â€˜/usr/local/lib/R/site-libraryâ€™
(as â€˜libâ€™ is unspecified)



[36mâ”€â”€[39m [36mR CMD build[39m [36mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€[39m
* checking for file â€˜/tmp/RtmpNviskA/remotesfa163c8a18/cole-trapnell-lab-monocle3-98402ed/DESCRIPTIONâ€™ ... OK
* preparing â€˜monocle3â€™:
* checking DESCRIPTION meta-information ... OK
* cleaning src
* checking for LF line-endings in source and make files and shell scripts
* checking for empty or unneeded directories
Omitted â€˜LazyDataâ€™ from DESCRIPTION
* building â€˜monocle3_1.3.7.tar.gzâ€™



Installing package into â€˜/usr/local/lib/R/site-libraryâ€™
(as â€˜libâ€™ is unspecified)
</code></pre>
<pre><code class="language-R">library(monocle3)
</code></pre>
<pre><code>Loading required package: Biobase

Loading required package: BiocGenerics


Attaching package: â€˜BiocGenericsâ€™


The following objects are masked from â€˜package:statsâ€™:

    IQR, mad, sd, var, xtabs


The following objects are masked from â€˜package:baseâ€™:

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,
    table, tapply, union, unique, unsplit, which.max, which.min


Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.


Loading required package: SingleCellExperiment

Loading required package: SummarizedExperiment

Loading required package: MatrixGenerics

Loading required package: matrixStats


Attaching package: â€˜matrixStatsâ€™


The following objects are masked from â€˜package:Biobaseâ€™:

    anyMissing, rowMedians



Attaching package: â€˜MatrixGenericsâ€™


The following objects are masked from â€˜package:matrixStatsâ€™:

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars


The following object is masked from â€˜package:Biobaseâ€™:

    rowMedians


Loading required package: GenomicRanges

Loading required package: stats4

Loading required package: S4Vectors


Attaching package: â€˜S4Vectorsâ€™


The following object is masked from â€˜package:utilsâ€™:

    findMatches


The following objects are masked from â€˜package:baseâ€™:

    expand.grid, I, unname


Loading required package: IRanges

Loading required package: GenomeInfoDb


Attaching package: â€˜monocle3â€™


The following objects are masked from â€˜package:Biobaseâ€™:

    exprs, fData, fData&lt;-, pData, pData&lt;-
</code></pre>
<pre><code class="language-R">system(&quot;tar zcvf R_lib_monocle3.tar.gz /usr/local/lib/R/site-library&quot;)
</code></pre>
              
            </div>
          </div><footer>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>Copyright &copy; 2025 Shaojun Xie</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/Bix4UMD/BIOI611_lab.git" class="fa fa-code-fork" style="color: #fcfcfc"> Bix4UMD/BIOI611_lab</a>
        </span>
    
    
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../javascripts/mathjax.js"></script>
      <script src="https://unpkg.com/mathjax@3/es5/tex-mml-chtml.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
