<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Shaojun Xie" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Downstream analysis of 10x scRNA-seq data for C. elegans data - Lab note for UMD BIOI611</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Downstream analysis of 10x scRNA-seq data for C. elegans data";
        var mkdocs_page_input_path = "BIOI611_scRNA_cele.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/yaml.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/rust.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> Lab note for UMD BIOI611
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption"><span class="caption-text">Get ready for the course</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../basic_linux/">Basic Linux</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Bulk RNA-seq</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../bulkRNAseq_lab/">Read QC and alignment</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Phred_FQ/">Phred score in FASTQ</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_introR/">Minimal R Introduction</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_DESeq2_analysis/">DEG analysis using DESeq2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_DAVID/">Pathway analysis using DAVID</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_ballgown_example/">DE analysis at isoform-leve using ballgown (example data)</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_bulkRNA_SE_ballgown/">DE analysis at isoform-leve using ballgown (C. ele data)</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_long_read_transcriptome/">Long read transcriptome</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">scRNA-seq</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_analysis_10x_dataset_PBMC/">Initial analysis of 10x scRNA-seq data for human PBMC using cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_scRNA/">Downstream analysis of 10x scRNA-seq data for human PBMC using Seurat</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_scRNA_seq_cele_cellranger/">Initial analysis of 10x scRNA-seq data for C. elegans using cellranger</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">Downstream analysis of 10x scRNA-seq data for C. elegans data</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#install-required-r-packages">Install required R packages</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#load-required-r-packages">Load required R packages</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#access-seurat-object">Access Seurat object</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#data-preprocessing">Data preprocessing</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#how-to-interpret-qc-plot">How to interpret QC plot</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#filter-out-potential-doublets-empty-droplets-and-dying-cells">Filter out potential doublets, empty droplets and dying cells</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#normalization-ccaling-of-the-data-and-linear-dimensional-reduction">Normalization, ccaling of the data and linear dimensional reduction</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#normalization">Normalization</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scaling-the-data">Scaling the data</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#perform-analysis-without-integration">Perform analysis without integration</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#perform-integration">Perform integration</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#save-the-seurat-object">Save the Seurat object</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#reference">Reference</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bioi611_monocle_cele/">Downstream analysis - trajectory analysis using Monocle3</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">Lab note for UMD BIOI611</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">scRNA-seq</li>
      <li class="breadcrumb-item active">Downstream analysis of 10x scRNA-seq data for C. elegans data</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/Bix4UMD/BIOI611_lab.git/edit/master/docs/BIOI611_scRNA_cele.md">Edit on Bix4UMD/BIOI611_lab</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <p><a href="https://colab.research.google.com/github/Bix4UMD/BIOI611_lab/blob/main/docs/BIOI611_scRNA_cele.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<h1 id="_1"></h1>
<p>The following note is designed to give you an overview of comparative analyses on complex cell types that are possible using the Seurat integration procedure.</p>
<h2 id="install-required-r-packages">Install required R packages</h2>
<pre><code class="language-R"># # Install the remotes package
# if (!requireNamespace(&quot;remotes&quot;, quietly = TRUE)) {
#   install.packages(&quot;remotes&quot;)
# }
# # Install Seurat
# if (!requireNamespace(&quot;Seurat&quot;, quietly = TRUE)) {
#     remotes::install_github(&quot;satijalab/seurat&quot;, &quot;seurat5&quot;, quiet = TRUE)
# }
# # Install BiocManager
# if (!require(&quot;BiocManager&quot;, quietly = TRUE))
#     install.packages(&quot;BiocManager&quot;)

# # Install SingleR package
# if (!require(&quot;hdf5r&quot;, quietly = TRUE)){
#     BiocManager::install(&quot;hdf5r&quot;)
# }
# # Install SingleR package
# if (!require(&quot;presto&quot;, quietly = TRUE)){
#     remotes::install_github(&quot;immunogenomics/presto&quot;)
# }
# # Install SingleR package
# if (!require(&quot;SingleR&quot;, quietly = TRUE)){
#     BiocManager::install(&quot;SingleR&quot;)
# }
# if (!require(&quot;celldex&quot;, quietly = TRUE)){
#     BiocManager::install(&quot;celldex&quot;)
# }
# if (!require(&quot;SingleCellExperiment&quot;, quietly = TRUE)){
#     BiocManager::install(&quot;SingleCellExperiment&quot;)
# }
# if (!require(&quot;scater&quot;, quietly = TRUE)){
#     BiocManager::install(&quot;scater&quot;)
# }
</code></pre>
<pre><code class="language-R">## Installing the R packages could take around 51 minutes
## To speed up this process, you can download the R lib files
## saved from a working Google Colab session
## https://drive.google.com/file/d/1eN-Raj3Hlnml68ACZQ36VQRGZxFclneL/view?usp=sharing
system(&quot;gdown 1eN-Raj3Hlnml68ACZQ36VQRGZxFclneL&quot;, intern = TRUE)
</code></pre>
<pre><code class="language-R">system(&quot;md5sum BIOI611_custom_R_lib_Nov_2025.tar.gz&quot;, intern = TRUE)
</code></pre>
<p><span style=white-space:pre-wrap>'fcfbd50c028468ec37d8654fda0eaec6  BIOI611_custom_R_lib_Nov_2025.tar.gz'</span></p>
<p><span style=white-space:pre-wrap>'fcfbd50c028468ec37d8654fda0eaec6  BIOI611_custom_R_lib_Nov_2025.tar.gz'</span></p>
<pre><code class="language-R">## required by scater package
system(&quot;apt-get install libx11-dev libcairo2-dev&quot;, intern = TRUE)
system(&quot;tar zxvf BIOI611_custom_R_lib_Nov_2025.tar.gz&quot;)
</code></pre>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'Reading package lists...'</li><li>'Building dependency tree...'</li><li>'Reading state information...'</li><li>'libcairo2-dev is already the newest version (1.16.0-5ubuntu2).'</li><li>'libx11-dev is already the newest version (2:1.7.5-1ubuntu0.3).'</li><li>'0 upgraded, 0 newly installed, 0 to remove and 41 not upgraded.'</li></ol>

<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'Reading package lists...'</li><li>'Building dependency tree...'</li><li>'Reading state information...'</li><li>'libcairo2-dev is already the newest version (1.16.0-5ubuntu2).'</li><li>'libx11-dev is already the newest version (2:1.7.5-1ubuntu0.3).'</li><li>'0 upgraded, 0 newly installed, 0 to remove and 41 not upgraded.'</li></ol>

<pre><code class="language-R">.libPaths(c(&quot;/content/usr/local/lib/R/site-library&quot;, .libPaths()))
</code></pre>
<pre><code class="language-R">.libPaths()
</code></pre>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'/content/usr/local/lib/R/site-library'</li><li>'/usr/local/lib/R/site-library'</li><li>'/usr/lib/R/site-library'</li><li>'/usr/lib/R/library'</li></ol>

<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'/content/usr/local/lib/R/site-library'</li><li>'/usr/local/lib/R/site-library'</li><li>'/usr/lib/R/site-library'</li><li>'/usr/lib/R/library'</li></ol>

<h2 id="load-required-r-packages">Load required R packages</h2>
<pre><code class="language-R">library(Seurat)
library(dplyr)
library(SingleR)
library(celldex)
library(scater)
library(SingleCellExperiment)
</code></pre>
<pre><code>Loading required package: SeuratObject

Loading required package: sp


Attaching package: ‚ÄòSeuratObject‚Äô


The following objects are masked from ‚Äòpackage:base‚Äô:

    intersect, t



Attaching package: ‚Äòdplyr‚Äô


The following objects are masked from ‚Äòpackage:stats‚Äô:

    filter, lag


The following objects are masked from ‚Äòpackage:base‚Äô:

    intersect, setdiff, setequal, union


Loading required package: SummarizedExperiment

Loading required package: MatrixGenerics

Loading required package: matrixStats


Attaching package: ‚ÄòmatrixStats‚Äô


The following object is masked from ‚Äòpackage:dplyr‚Äô:

    count



Attaching package: ‚ÄòMatrixGenerics‚Äô


The following objects are masked from ‚Äòpackage:matrixStats‚Äô:

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars


Loading required package: GenomicRanges

Loading required package: stats4

Loading required package: BiocGenerics

Loading required package: generics


Attaching package: ‚Äògenerics‚Äô


The following object is masked from ‚Äòpackage:dplyr‚Äô:

    explain


The following objects are masked from ‚Äòpackage:base‚Äô:

    as.difftime, as.factor, as.ordered, intersect, is.element, setdiff,
    setequal, union



Attaching package: ‚ÄòBiocGenerics‚Äô


The following object is masked from ‚Äòpackage:dplyr‚Äô:

    combine


The following objects are masked from ‚Äòpackage:stats‚Äô:

    IQR, mad, sd, var, xtabs


The following objects are masked from ‚Äòpackage:base‚Äô:

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, is.unsorted, lapply, Map, mapply, match, mget,
    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
    rbind, Reduce, rownames, sapply, saveRDS, table, tapply, unique,
    unsplit, which.max, which.min


Loading required package: S4Vectors


Attaching package: ‚ÄòS4Vectors‚Äô


The following objects are masked from ‚Äòpackage:dplyr‚Äô:

    first, rename


The following object is masked from ‚Äòpackage:utils‚Äô:

    findMatches


The following objects are masked from ‚Äòpackage:base‚Äô:

    expand.grid, I, unname


Loading required package: IRanges


Attaching package: ‚ÄòIRanges‚Äô


The following objects are masked from ‚Äòpackage:dplyr‚Äô:

    collapse, desc, slice


The following object is masked from ‚Äòpackage:sp‚Äô:

    %over%


Loading required package: Seqinfo

Loading required package: Biobase

Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.



Attaching package: ‚ÄòBiobase‚Äô


The following object is masked from ‚Äòpackage:MatrixGenerics‚Äô:

    rowMedians


The following objects are masked from ‚Äòpackage:matrixStats‚Äô:

    anyMissing, rowMedians



Attaching package: ‚ÄòSummarizedExperiment‚Äô


The following object is masked from ‚Äòpackage:Seurat‚Äô:

    Assays


The following object is masked from ‚Äòpackage:SeuratObject‚Äô:

    Assays



Attaching package: ‚Äòcelldex‚Äô


The following objects are masked from ‚Äòpackage:SingleR‚Äô:

    BlueprintEncodeData, DatabaseImmuneCellExpressionData,
    HumanPrimaryCellAtlasData, ImmGenData, MonacoImmuneData,
    MouseRNAseqData, NovershternHematopoieticData


Loading required package: SingleCellExperiment

Loading required package: scuttle

Loading required package: ggplot2
</code></pre>
<pre><code class="language-R">list.files()
</code></pre>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'BIOI611_custom_R_lib_Nov_2025.tar.gz'</li><li>'cele_cellranger_mtx'</li><li>'sample_data'</li><li>'sessionInfo.txt'</li><li>'usr'</li></ol>

<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'BIOI611_custom_R_lib_Nov_2025.tar.gz'</li><li>'cele_cellranger_mtx'</li><li>'sample_data'</li><li>'sessionInfo.txt'</li><li>'Seurat_object_10x_cele_final.rds'</li><li>'Seurat_object_10x_cele_merged_integ.cloupe'</li><li>'usr'</li></ol>

<pre><code class="language-R"># https://drive.google.com/drive/folders/1lp6kSGFyYYAswfAyG07DgELQ2G2Ja51Q?usp=sharing
# Download &quot;filtered_feature_bc_matrix.h5&quot;
# Output of cellranger
system(&quot;gdown --folder 1lp6kSGFyYYAswfAyG07DgELQ2G2Ja51Q&quot;, intern = TRUE)
</code></pre>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'Processing file 1ezH8P0iRpV9fsOOqrqStPxI5-q54Ge9B filtered_feature_bc_matrix_300min.h5'</li><li>'Processing file 1hu9c2BhKb5bEYXrPlG_mkSQ219eG-E2K filtered_feature_bc_matrix_400min.h5'</li><li>'Processing file 16bSSWAn5Fg_sZgajA4JbuZPoVgulGt_5 filtered_feature_bc_matrix_500min.h5'</li></ol>

<pre><code class="language-R">system(&quot;md5sum ./cele_cellranger_mtx/*.h5&quot;, intern = TRUE)
</code></pre>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li><span style=white-space:pre-wrap>'e0fd344696c5188e55aeb359efd7a8c1  ./cele_cellranger_mtx/filtered_feature_bc_matrix_300min.h5'</span></li><li><span style=white-space:pre-wrap>'d087ff62ba449586858c058117aa0438  ./cele_cellranger_mtx/filtered_feature_bc_matrix_400min.h5'</span></li><li><span style=white-space:pre-wrap>'efb8a9ef4898918e53a53878531f64ce  ./cele_cellranger_mtx/filtered_feature_bc_matrix_500min.h5'</span></li></ol>

<pre><code class="language-R"># Specify the directory containing the .h5 files
mtx_directory &lt;- &quot;./cele_cellranger_mtx&quot;

# List all .h5 files in the specified directory
mtx_file_paths &lt;- list.files(path = mtx_directory, pattern = &quot;\\.h5$&quot;, full.names = TRUE)

# Print the file paths
mtx_file_paths
</code></pre>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'./cele_cellranger_mtx/filtered_feature_bc_matrix_300min.h5'</li><li>'./cele_cellranger_mtx/filtered_feature_bc_matrix_400min.h5'</li><li>'./cele_cellranger_mtx/filtered_feature_bc_matrix_500min.h5'</li></ol>

<pre><code class="language-R"># Read the files into a list
count_mtx_list &lt;- lapply(mtx_file_paths, Read10X_h5)
</code></pre>
<pre><code class="language-R">names(count_mtx_list) &lt;- c(&quot;300min&quot;, &quot;400min&quot;, &quot;500min&quot;)
names(count_mtx_list)
</code></pre>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'300min'</li><li>'400min'</li><li>'500min'</li></ol>

<pre><code class="language-R">lapply(count_mtx_list, class)
</code></pre>
<dl>
    <dt>$`300min`</dt>
        <dd>'dgCMatrix'</dd>
    <dt>$`400min`</dt>
        <dd>'dgCMatrix'</dd>
    <dt>$`500min`</dt>
        <dd>'dgCMatrix'</dd>
</dl>

<pre><code class="language-R">print(format(object.size(count_mtx_list), units = &quot;MB&quot;))
</code></pre>
<pre><code>[1] "829.8 Mb"
</code></pre>
<pre><code class="language-R">sample_names &lt;- names(count_mtx_list)

</code></pre>
<pre><code class="language-R">seurat_obj_list &lt;- list()
for (i in seq_along(count_mtx_list)) {
  seurat_obj &lt;- CreateSeuratObject(counts = count_mtx_list[[i]],
                                   project = sample_names[i])
  seurat_obj_list[[sample_names[i]]] &lt;- seurat_obj
}
</code></pre>
<pre><code>Warning message:
‚ÄúFeature names cannot have underscores ('_'), replacing with dashes ('-')‚Äù
Warning message:
‚ÄúFeature names cannot have underscores ('_'), replacing with dashes ('-')‚Äù
Warning message:
‚ÄúFeature names cannot have underscores ('_'), replacing with dashes ('-')‚Äù
</code></pre>
<pre><code class="language-R">seurat_obj_list
</code></pre>
<pre><code>$`300min`
An object of class Seurat 
19985 features across 25996 samples within 1 assay 
Active assay: RNA (19985 features, 0 variable features)
 1 layer present: counts

$`400min`
An object of class Seurat 
19985 features across 37944 samples within 1 assay 
Active assay: RNA (19985 features, 0 variable features)
 1 layer present: counts

$`500min`
An object of class Seurat 
19985 features across 14378 samples within 1 assay 
Active assay: RNA (19985 features, 0 variable features)
 1 layer present: counts
</code></pre>
<pre><code class="language-R">rm(count_mtx_list); gc();
</code></pre>
<table class="dataframe">
<caption>A matrix: 2 √ó 6 of type dbl</caption>
<thead>
    <tr><th></th><th scope=col>used</th><th scope=col>(Mb)</th><th scope=col>gc trigger</th><th scope=col>(Mb)</th><th scope=col>max used</th><th scope=col>(Mb)</th></tr>
</thead>
<tbody>
    <tr><th scope=row>Ncells</th><td> 11389606</td><td>608.3</td><td> 20298022</td><td>1084.1</td><td> 16216372</td><td> 866.1</td></tr>
    <tr><th scope=row>Vcells</th><td>128158737</td><td>977.8</td><td>322702913</td><td>2462.1</td><td>322696490</td><td>2462.0</td></tr>
</tbody>
</table>

<h3 id="access-seurat-object">Access Seurat object</h3>
<pre><code class="language-R">seurat_obj_list$'300min'@active.assay
</code></pre>
<p>'RNA'</p>
<pre><code class="language-R">class(seurat_obj_list$'300min'@meta.data)
head(seurat_obj_list$'300min'@meta.data, 4)
</code></pre>
<p>'data.frame'</p>
<table class="dataframe">
<caption>A data.frame: 4 √ó 3</caption>
<thead>
    <tr><th></th><th scope=col>orig.ident</th><th scope=col>nCount_RNA</th><th scope=col>nFeature_RNA</th></tr>
    <tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>
</thead>
<tbody>
    <tr><th scope=row>AAACCTGAGACAATAC-1</th><td>300min</td><td>1630</td><td> 803</td></tr>
    <tr><th scope=row>AAACCTGAGACACTAA-1</th><td>300min</td><td>3147</td><td>1365</td></tr>
    <tr><th scope=row>AAACCTGAGACGCTTT-1</th><td>300min</td><td> 892</td><td> 586</td></tr>
    <tr><th scope=row>AAACCTGAGAGGGCTT-1</th><td>300min</td><td>1666</td><td>1033</td></tr>
</tbody>
</table>

<pre><code class="language-R">seurat_obj_list$'300min'@meta.data$orig.ident = &quot;300min&quot;
seurat_obj_list$'400min'@meta.data$orig.ident = &quot;400min&quot;
seurat_obj_list$'500min'@meta.data$orig.ident = &quot;500min&quot;
</code></pre>
<pre><code class="language-R">head(seurat_obj_list$'300min'@meta.data, 4)
</code></pre>
<table class="dataframe">
<caption>A data.frame: 4 √ó 3</caption>
<thead>
    <tr><th></th><th scope=col>orig.ident</th><th scope=col>nCount_RNA</th><th scope=col>nFeature_RNA</th></tr>
    <tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>
</thead>
<tbody>
    <tr><th scope=row>AAACCTGAGACAATAC-1</th><td>300min</td><td>1630</td><td> 803</td></tr>
    <tr><th scope=row>AAACCTGAGACACTAA-1</th><td>300min</td><td>3147</td><td>1365</td></tr>
    <tr><th scope=row>AAACCTGAGACGCTTT-1</th><td>300min</td><td> 892</td><td> 586</td></tr>
    <tr><th scope=row>AAACCTGAGAGGGCTT-1</th><td>300min</td><td>1666</td><td>1033</td></tr>
</tbody>
</table>

<pre><code class="language-R">Layers(seurat_obj_list$'300min')
</code></pre>
<p>'counts'</p>
<pre><code class="language-R">seurat_obj_list$'300min'@version
</code></pre>
<pre><code>[1] ‚Äò5.2.0‚Äô
</code></pre>
<h2 id="data-preprocessing">Data preprocessing</h2>
<p>Ensembl biomart can be used to extract the mitochodria genes:</p>
<p>https://useast.ensembl.org/info/website/archives/assembly.html</p>
<table>
<thead>
<tr>
<th>Gene stable ID</th>
<th>Gene name</th>
</tr>
</thead>
<tbody>
<tr>
<td>WBGene00000829</td>
<td>ctb-1</td>
</tr>
<tr>
<td>WBGene00010957</td>
<td>nduo-6</td>
</tr>
<tr>
<td>WBGene00010958</td>
<td>WBGene00010958</td>
</tr>
<tr>
<td>WBGene00010959</td>
<td>WBGene00010959</td>
</tr>
<tr>
<td>WBGene00010960</td>
<td>atp-6</td>
</tr>
<tr>
<td>WBGene00010961</td>
<td>nduo-2</td>
</tr>
<tr>
<td>WBGene00010962</td>
<td>ctc-3</td>
</tr>
<tr>
<td>WBGene00010963</td>
<td>nduo-4</td>
</tr>
<tr>
<td>WBGene00010964</td>
<td>ctc-1</td>
</tr>
<tr>
<td>WBGene00010965</td>
<td>ctc-2</td>
</tr>
<tr>
<td>WBGene00010966</td>
<td>nduo-3</td>
</tr>
<tr>
<td>WBGene00010967</td>
<td>nduo-5</td>
</tr>
</tbody>
</table>
<pre><code class="language-R"># Define the mitochondria gene names as an R vector
mt_gene_names &lt;- c(
  &quot;ctb-1&quot;, &quot;nduo-6&quot;, &quot;WBGene00010958&quot;, &quot;WBGene00010959&quot;,
  &quot;atp-6&quot;, &quot;nduo-2&quot;, &quot;ctc-3&quot;, &quot;nduo-4&quot;,
  &quot;ctc-1&quot;, &quot;ctc-2&quot;, &quot;nduo-3&quot;, &quot;nduo-5&quot;
)
mt_genes &lt;- mt_gene_names[mt_gene_names %in% rownames(seurat_obj_list$'300min')]
mt_genes
</code></pre>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'ctb-1'</li><li>'nduo-6'</li><li>'WBGene00010958'</li><li>'WBGene00010959'</li><li>'atp-6'</li><li>'nduo-2'</li><li>'ctc-3'</li><li>'nduo-4'</li><li>'ctc-1'</li><li>'ctc-2'</li><li>'nduo-3'</li><li>'nduo-5'</li></ol>

<pre><code class="language-R"># Function to calculate percentage of mitochondrial genes
add_mt_percentage &lt;- function(seurat_obj, mt_genes) {
  # Calculate percentage of mitochondrial genes
  seurat_obj$percent.mt &lt;- PercentageFeatureSet(seurat_obj, features = mt_genes)
  return(seurat_obj)
}
</code></pre>
<pre><code class="language-R">seurat_obj_list &lt;- lapply(seurat_obj_list, add_mt_percentage, mt_genes = mt_genes)
</code></pre>
<pre><code class="language-R">qc_features &lt;- c(&quot;nFeature_RNA&quot;, &quot;nCount_RNA&quot;, &quot;percent.mt&quot;)
</code></pre>
<pre><code class="language-R">VlnPlot(seurat_obj_list$'300min', features = qc_features, ncol = 3, pt.size=0)
VlnPlot(seurat_obj_list$'400min', features = qc_features, ncol = 3, pt.size=0)
VlnPlot(seurat_obj_list$'500min', features = qc_features, ncol = 3, pt.size=0)
</code></pre>
<pre><code>Warning message:
‚ÄúDefault search for "data" layer in "RNA" assay yielded no results; utilizing "counts" layer instead.‚Äù
Warning message:
‚Äú[1m[22mThe `slot` argument of `FetchData()` is deprecated as of SeuratObject 5.0.0.
[36m‚Ñπ[39m Please use the `layer` argument instead.
[36m‚Ñπ[39m The deprecated feature was likely used in the [34mSeurat[39m package.
  Please report the issue at [3m[34m&lt;https://github.com/satijalab/seurat/issues&gt;[39m[23m.‚Äù
Warning message:
‚Äú[1m[22m`PackageCheck()` was deprecated in SeuratObject 5.0.0.
[36m‚Ñπ[39m Please use `rlang::check_installed()` instead.
[36m‚Ñπ[39m The deprecated feature was likely used in the [34mSeurat[39m package.
  Please report the issue at [3m[34m&lt;https://github.com/satijalab/seurat/issues&gt;[39m[23m.‚Äù
Warning message:
‚Äú[1m[22m`aes_string()` was deprecated in ggplot2 3.0.0.
[36m‚Ñπ[39m Please use tidy evaluation idioms with `aes()`.
[36m‚Ñπ[39m See also `vignette("ggplot2-in-packages")` for more information.
[36m‚Ñπ[39m The deprecated feature was likely used in the [34mSeurat[39m package.
  Please report the issue at [3m[34m&lt;https://github.com/satijalab/seurat/issues&gt;[39m[23m.‚Äù
Warning message:
‚ÄúDefault search for "data" layer in "RNA" assay yielded no results; utilizing "counts" layer instead.‚Äù
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_cele_files/BIOI611_scRNA_cele_36_1.png" /></p>
<pre><code>Warning message:
‚ÄúDefault search for "data" layer in "RNA" assay yielded no results; utilizing "counts" layer instead.‚Äù
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_cele_files/BIOI611_scRNA_cele_36_3.png" /></p>
<p><img alt="png" src="../BIOI611_scRNA_cele_files/BIOI611_scRNA_cele_36_4.png" /></p>
<h3 id="how-to-interpret-qc-plot">How to interpret QC plot</h3>
<p><code>nFeature_RNA</code>: The number of unique features (genes) detected per cell.</p>
<p>Extremely high values could suggest potential doublets (two cells mistakenly captured as one), as two cells would have more unique genes combined.</p>
<p>Low number of detected genes - potential ambient mRNA (not real cells)</p>
<p><code>nCount_RNA</code>: The total number of RNA molecules (or unique molecular identifiers, UMIs) detected per cell.</p>
<p>Higher counts generally indicate higher RNA content, but they could also result from cell doublets.
Cells with very low nCount_RNA might represent poor-quality cells with low RNA capture, while very high counts may also suggest doublets.</p>
<p><code>percent.mt</code>: The percentage of reads mapping to mitochondrial genes.</p>
<p>High mitochondrial content often indicates cell stress or apoptosis, as damaged cells tend to release mitochondrial RNA.</p>
<p>Filtering cells with high <code>percent.mt</code> values is common to exclude potentially dying cells.</p>
<pre><code class="language-R"># FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 &lt;- FeatureScatter(seurat_obj_list$'300min', feature1 = &quot;nCount_RNA&quot;, feature2 = &quot;percent.mt&quot;)
plot2 &lt;- FeatureScatter(seurat_obj_list$'300min', feature1 = &quot;nCount_RNA&quot;, feature2 = &quot;nFeature_RNA&quot;)
plot1 + plot2
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_cele_files/BIOI611_scRNA_cele_38_0.png" /></p>
<h3 id="filter-out-potential-doublets-empty-droplets-and-dying-cells">Filter out potential doublets, empty droplets and dying cells</h3>
<pre><code class="language-R"># Load necessary libraries
library(Seurat)
library(ggplot2)

# Define the function to calculate median and MAD values
calculate_thresholds &lt;- function(seurat_obj) {
  # Extract relevant columns
  nFeature_values &lt;- seurat_obj@meta.data$nFeature_RNA
  nCount_values &lt;- seurat_obj@meta.data$nCount_RNA
  percent_mt_values &lt;- seurat_obj@meta.data$percent.mt

  # Calculate medians and MADs
  nFeature_median &lt;- median(nFeature_values, na.rm = TRUE)
  nFeature_mad &lt;- mad(nFeature_values, constant = 1, na.rm = TRUE)

  nCount_median &lt;- median(nCount_values, na.rm = TRUE)
  nCount_mad &lt;- mad(nCount_values, constant = 1, na.rm = TRUE)

  percent_mt_median &lt;- median(percent_mt_values, na.rm = TRUE)
  percent_mt_mad &lt;- mad(percent_mt_values, constant = 1, na.rm = TRUE)

  # Calculate thresholds for horizontal lines
  thresholds &lt;- list(
    nFeature_upper = nFeature_median + 4 * nFeature_mad,
    nFeature_lower = nFeature_median - 4 * nFeature_mad,
    nCount_upper = nCount_median + 4 * nCount_mad,
    nCount_lower = nCount_median - 4 * nCount_mad,
    percent_mt_upper = percent_mt_median + 4 * percent_mt_mad
  )

  return(thresholds)
}

# Define a function to filter Seurat objects
filter_seurat_obj &lt;- function(seurat_obj) {
  # Calculate thresholds
  thresholds &lt;- calculate_thresholds(seurat_obj)

  # Apply filtering
  seurat_obj &lt;- subset(
    seurat_obj,
    subset = nFeature_RNA &gt; thresholds$nFeature_lower &amp;
             nFeature_RNA &lt; thresholds$nFeature_upper &amp;
             percent.mt &lt; thresholds$percent_mt_upper
  )
  #
  return(seurat_obj)
}

</code></pre>
<pre><code class="language-R"># Apply filtering to each Seurat object in the list
seurat_obj_list &lt;- lapply(seurat_obj_list, filter_seurat_obj)
</code></pre>
<pre><code class="language-R">
</code></pre>
<pre><code class="language-R">
</code></pre>
<pre><code class="language-R">VlnPlot(seurat_obj_list$'300min', features = qc_features, ncol = 3, pt.size=0)
VlnPlot(seurat_obj_list$'400min', features = qc_features, ncol = 3, pt.size=0)
VlnPlot(seurat_obj_list$'500min', features = qc_features, ncol = 3, pt.size=0)
</code></pre>
<pre><code>Warning message:
‚ÄúDefault search for "data" layer in "RNA" assay yielded no results; utilizing "counts" layer instead.‚Äù
Warning message:
‚ÄúDefault search for "data" layer in "RNA" assay yielded no results; utilizing "counts" layer instead.‚Äù
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_cele_files/BIOI611_scRNA_cele_44_1.png" /></p>
<pre><code>Warning message:
‚ÄúDefault search for "data" layer in "RNA" assay yielded no results; utilizing "counts" layer instead.‚Äù
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_cele_files/BIOI611_scRNA_cele_44_3.png" /></p>
<p><img alt="png" src="../BIOI611_scRNA_cele_files/BIOI611_scRNA_cele_44_4.png" /></p>
<pre><code class="language-R">RandomSubsetSeurat &lt;- function(seurat_obj, subset_size, seed = NULL) {
  # Optionally set a random seed for reproducibility
  if (!is.null(seed)) {
    set.seed(seed)
  }

  # Get all cell names
  total_cells &lt;- Cells(seurat_obj)

  # Ensure subset size is not larger than the total number of cells
  if (subset_size &gt; length(total_cells)) {
    stop(&quot;Subset size exceeds the total number of cells in the Seurat object.&quot;)
  }

  # Randomly sample a subset of cell names
  subset_cells &lt;- sample(total_cells, size = subset_size)

  # Create a new Seurat object with the subsetted cells
  subset_seurat_obj &lt;- subset(seurat_obj, cells = subset_cells)

  return(subset_seurat_obj)
}

</code></pre>
<p>‚ö†Ô∏è Important</p>
<p>This is included only to reduce the memory used. In real project, you don't want to perform this step. Instead, you should request larger memory computing resources.</p>
<pre><code class="language-R">seurat_obj_list &lt;- lapply(seurat_obj_list,
                      RandomSubsetSeurat,
                      subset_size = 2000,
                      seed = 123)
</code></pre>
<pre><code class="language-R">seurat_obj_list
</code></pre>
<pre><code>$`300min`
An object of class Seurat 
19985 features across 2000 samples within 1 assay 
Active assay: RNA (19985 features, 0 variable features)
 1 layer present: counts

$`400min`
An object of class Seurat 
19985 features across 2000 samples within 1 assay 
Active assay: RNA (19985 features, 0 variable features)
 1 layer present: counts

$`500min`
An object of class Seurat 
19985 features across 2000 samples within 1 assay 
Active assay: RNA (19985 features, 0 variable features)
 1 layer present: counts
</code></pre>
<pre><code class="language-R">so_merged &lt;- merge(seurat_obj_list$'300min',
         c(seurat_obj_list$'400min', seurat_obj_list$'500min'),
         add.cell.ids = c(&quot;300min&quot;, &quot;400min&quot;, &quot;500min&quot;),
         project = &quot;scRNA_cele&quot;)
so_merged
Layers(so_merged)
table(so_merged$orig.ident)
</code></pre>
<pre><code>An object of class Seurat 
19985 features across 6000 samples within 1 assay 
Active assay: RNA (19985 features, 0 variable features)
 3 layers present: counts.300min, counts.400min, counts.500min
</code></pre>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'counts.300min'</li><li>'counts.400min'</li><li>'counts.500min'</li></ol>

<pre><code>300min 400min 500min 
  2000   2000   2000
</code></pre>
<pre><code class="language-R">#rm(seurat_obj_list); gc();

</code></pre>
<p>Instead of using an arbitrary number, you can also use statistical algorithm to predict doublets and empty droplets to filter the cells, such as <code>DoubletFinder</code> and <code>EmptyDrops</code>.</p>
<h3 id="normalization-ccaling-of-the-data-and-linear-dimensional-reduction">Normalization, ccaling of the data and linear dimensional reduction</h3>
<h3 id="normalization">Normalization</h3>
<p>After removing unwanted cells from the dataset, the next step is to normalize the data. By default, a global-scaling normalization method ‚ÄúLogNormalize‚Äù that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. In Seurat v5, Normalized values are stored in <code>pbmc[["RNA"]]$data</code>.</p>
<p>While this method of normalization is standard and widely used in scRNA-seq analysis, global-scaling relies on an assumption that each cell originally contains the same number of RNA molecules.</p>
<p>Next, we identify a subset of features that show high variation across cells in the dataset‚Äîmeaning they are highly expressed in some cells and lowly expressed in others. Prior work, including our own, has shown that focusing on these variable genes in downstream analyses can enhance the detection of biological signals in single-cell datasets.</p>
<p>The approach used in Seurat improves upon previous versions by directly modeling the inherent mean-variance relationship in single-cell data. This method is implemented in the FindVariableFeatures() function, which, by default, selects 2,000 variable features per dataset. These features will then be used in downstream analyses, such as PCA.</p>
<pre><code class="language-R">
</code></pre>
<h3 id="scaling-the-data">Scaling the data</h3>
<p>Next, we apply a linear transformation (<code>scaling</code>) that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The <code>ScaleData()</code> function:</p>
<p>Shifts the expression of each gene, so that the mean expression across cells is 0
Scales the expression of each gene, so that the variance across cells is 1</p>
<p>This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate</p>
<p>The results of this are stored in <code>pbmc[["RNA"]]$scale.data</code></p>
<p>By default, only variable features are scaled.
You can specify the features argument to scale additional features.</p>
<pre><code class="language-R"># run standard anlaysis workflow
so_merged &lt;- NormalizeData(so_merged)
so_merged &lt;- FindVariableFeatures(so_merged)
so_merged &lt;- ScaleData(so_merged)
so_merged &lt;- RunPCA(so_merged)
</code></pre>
<pre><code>Normalizing layer: counts.300min

Normalizing layer: counts.400min

Normalizing layer: counts.500min

Finding variable features for layer counts.300min

Finding variable features for layer counts.400min

Finding variable features for layer counts.500min

Centering and scaling data matrix

PC_ 1 
Positive:  ost-1, pat-10, D2092.4, mlc-3, let-2, unc-15, lev-11, emb-9, tni-1, set-18 
       tnt-3, sgn-1, test-1, hsp-12.1, unc-98, cpn-3, sgcb-1, F53F10.1, spp-15, mup-2 
       Y71F9AR.2, mlc-1, C29F5.1, mlc-2, clik-1, stn-2, mig-18, F21H7.3, unc-60, Y73F8A.26 
Negative:  noah-1, noah-2, dpy-2, dpy-3, mlt-11, col-76, mlt-8, dpy-7, dpy-10, hch-1 
       col-121, dpy-17, dpy-14, txdc-12.2, sym-1, C01H6.8, Y41D4B.6, sqt-3, Y23H5B.8, K02E10.4 
       acn-1, C05C8.7, C26B9.3, R148.5, C48E7.1, dsl-6, inx-12, cpg-24, R05D3.9, F37C4.4 
PC_ 2 
Positive:  asp-4, enpl-1, C50F4.6, T02E9.5, his-24, atz-1, R07E5.17, C03C10.5, nphp-1, hil-3 
       tmem-231, mksr-2, tctn-1, fmi-1, jbts-14, osm-5, bbs-9, ift-81, fbxb-66, K02B12.2 
       mks-2, ift-20, K07C11.10, che-13, R01H2.8, ifta-2, F48E3.9, arl-3, ccep-290, tmem-17 
Negative:  unc-15, sgn-1, D2092.4, C29F5.1, let-2, lev-11, tnt-3, mlc-2, mup-2, mlc-1 
       tni-1, test-1, mig-18, clik-1, hsp-12.1, mlc-3, ttn-1, F21H7.3, sgca-1, emb-9 
       set-18, icl-1, ost-1, myo-3, unc-54, sgcb-1, hsp-12.2, unc-98, stn-2, Y71F9AR.2 
PC_ 3 
Positive:  mks-2, arl-3, mksr-2, ift-81, ifta-2, dylt-2, tmem-231, K07C11.10, che-13, ift-20 
       ccep-290, bbs-5, osm-5, C33A12.4, ift-74, R01H2.8, Y102A11A.9, bbs-9, tctn-1, lgc-20 
       dyf-1, mks-1, T02G5.3, dyf-3, arl-13, nphp-1, tmem-17, Y17D7B.10, bbs-2, F01E11.3 
Negative:  his-24, Y37E3.30, atz-1, lbp-1, C01G6.3, cht-1, ttr-50, clec-266, Y65A5A.1, enpl-1 
       hil-3, clec-196, idh-1, dsl-3, fbxb-66, cpg-20, F53B3.5, Y71F9AL.7, fbn-1, E01G4.5 
       pmt-2, ZK512.1, cutl-2, Y43F8B.2, cpg-24, F55C9.5, Y71F9AL.6, W04H10.6, fbxb-101, lam-3 
PC_ 4 
Positive:  arl-3, ifta-2, mks-2, mksr-2, tmem-231, ift-81, che-13, K07C11.10, ift-20, dylt-2 
       Y55D5A.1, bbs-9, ift-74, ccep-290, tctn-1, dsl-3, cutl-2, T02G5.3, R01H2.8, bbs-5 
       osm-5, nphp-1, dyf-1, lgc-20, dyf-3, arl-13, mks-1, Y102A11A.9, tmem-17, bbs-2 
Negative:  mab-7, wrt-2, abu-13, mlt-9, sups-1, clec-180, Y73E7A.8, C01G6.9, wrt-1, R07E3.6 
       Y54G2A.76, K08B12.1, mam-3, T19A5.3, glf-1, ZC449.1, H03E18.1, M03B6.3, ZK154.1, Y11D7A.9 
       pqn-32, F01D5.6, C35A5.11, F33D4.6, C52G5.2, grl-15, T03D8.6, F23H12.5, cut-6, ZC123.1 
PC_ 5 
Positive:  Y41D4B.6, B0205.4, C06C3.4, T01D1.8, F43D9.1, F11E6.9, F31C3.6, dsl-6, ttr-2, T19C4.1 
       F46G11.6, K02E10.4, best-14, nhx-1, ifa-3, T25B9.1, F46F11.7, Y45F10B.59, F15B9.8, C24B5.4 
       T19B10.5, clec-78, cpt-4, C49F5.13, srm-1, H41C03.1, B0393.5, C53A5.2, ttr-3, far-5 
Negative:  F33H2.8, nhr-127, F37A4.3, bus-17, nhr-270, sams-1, bus-12, F18A11.2, W04H10.6, oac-51 
       elt-1, F54B11.10, T13H10.2, nhr-218, rocf-1, Y6D1A.2, txdc-12.1, T04G9.4, F32H2.6, subs-4 
       fbn-1, F13G3.3, T03G6.1, nhr-94, C35A5.5, fbxa-52, fasn-1, nstp-3, W03B1.3, bus-8
</code></pre>
<pre><code class="language-R">DimPlot(so_merged,
        reduction = &quot;pca&quot;,
        split.by = 'orig.ident',
        label.color = &quot;black&quot;) +
        NoLegend()
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_cele_files/BIOI611_scRNA_cele_60_0.png" /></p>
<pre><code class="language-R">so_merged &lt;- FindNeighbors(so_merged,
                           dims = 1:30, reduction = &quot;pca&quot;)
so_merged &lt;- FindClusters(so_merged,
                           resolution = 2,
                           cluster.name = &quot;unintegrated_clusters&quot;)
</code></pre>
<pre><code>Computing nearest neighbor graph

Computing SNN



Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 6000
Number of edges: 197290

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8357
Number of communities: 37
Elapsed time: 0 seconds
</code></pre>
<p>You have several useful ways to visualize both cells and features that define the PCA, including <code>VizDimReduction()</code>, <code>DimPlot()</code>, and <code>DimHeatmap()</code>.</p>
<p><code>DimHeatmap()</code> draws a heatmap focusing on a principal component. Both cells and genes are sorted by their principal component scores</p>
<h2 id="perform-analysis-without-integration">Perform analysis without integration</h2>
<p>To visualize and explore these datasets, Seurat  offers several non-linear dimensional reduction techniques, such as tSNE and UMAP.</p>
<p>The goal of tSNE/UMAP is to learn underlying structure in the dataset, in order to place similar cells together in low-dimensional space. Therefore, cells that are grouped together within graph-based clusters determined above should co-localize on these dimension reduction plots.</p>
<pre><code class="language-R">so_merged &lt;- RunUMAP(so_merged,
              dims = 1:30,
              reduction = &quot;pca&quot;,
              reduction.name = &quot;umap.unintegrated&quot;)
</code></pre>
<pre><code>Warning message:
‚ÄúThe default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
This message will be shown once per session‚Äù
11:40:29 UMAP embedding parameters a = 0.9922 b = 1.112

11:40:29 Read 6000 rows and found 30 numeric columns

11:40:29 Using Annoy for neighbor search, n_neighbors = 30

11:40:29 Building Annoy index with metric = cosine, n_trees = 50

0%   10   20   30   40   50   60   70   80   90   100%

[----|----|----|----|----|----|----|----|----|----|

*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
|

11:40:30 Writing NN index file to temp file /tmp/RtmpCzdl7E/file783139c73f6

11:40:30 Searching Annoy index using 1 thread, search_k = 3000

11:40:32 Annoy recall = 100%

11:40:34 Commencing smooth kNN distance calibration using 1 thread
 with target n_neighbors = 30

11:40:36 Initializing from normalized Laplacian + noise (using RSpectra)

11:40:36 Commencing optimization for 500 epochs, with 241186 positive edges

11:40:36 Using rng type: pcg

11:40:46 Optimization finished
</code></pre>
<pre><code class="language-R">DimPlot(so_merged,
             reduction = &quot;umap.unintegrated&quot;,
             split.by = c(&quot;orig.ident&quot;))
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_cele_files/BIOI611_scRNA_cele_66_0.png" /></p>
<pre><code class="language-R">FeaturePlot(so_merged, feature=&quot;ham-1&quot;, pt.size = 0.1,
             split.by = c(&quot;orig.ident&quot;))
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_cele_files/BIOI611_scRNA_cele_67_0.png" /></p>
<pre><code class="language-R">FeaturePlot(so_merged, feature=&quot;egl-21&quot;, pt.size = 0.1,
             split.by = c(&quot;orig.ident&quot;))
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_cele_files/BIOI611_scRNA_cele_68_0.png" /></p>
<pre><code class="language-R">FeaturePlot(so_merged, feature=&quot;dsl-3&quot;, pt.size = 0.1,
             split.by = c(&quot;orig.ident&quot;))
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_cele_files/BIOI611_scRNA_cele_69_0.png" /></p>
<h2 id="perform-integration">Perform integration</h2>
<p>Seurat v5 enables streamlined integrative analysis using the IntegrateLayers function. The method currently supports five integration methods. Each of these methods performs integration in low-dimensional space, and returns a dimensional reduction (i.e. integrated.rpca) that aims to co-embed shared cell types across batches:</p>
<ul>
<li>Anchor-based CCA integration (method=CCAIntegration)</li>
<li>Harmony (method=HarmonyIntegration)</li>
<li>Anchor-based RPCA integration (method=RPCAIntegration)</li>
<li>FastMNN (method= FastMNNIntegration)</li>
<li>scVI (method=scVIIntegration)</li>
</ul>
<p>Canonical correlation analysis: CCA</p>
<p>Reciprocal PCA: RPCA</p>
<p><code>CCAIntegration</code> integration method that is available in the Seurat package utilizes the canonical correlation analysis (CCA). This method expects ‚Äúcorrespondences‚Äù or shared biological states among at least a subset of single cells across the groups.</p>
<pre><code class="language-R">so_merged_integ &lt;- IntegrateLayers(object = so_merged,
                method = CCAIntegration, orig.reduction = &quot;pca&quot;,
                new.reduction = &quot;integrated.cca&quot;,
                verbose = FALSE)

</code></pre>
<pre><code>Warning message:
‚Äú[1m[22mThe `slot` argument of `GetAssayData()` is deprecated as of SeuratObject 5.0.0.
[36m‚Ñπ[39m Please use the `layer` argument instead.
[36m‚Ñπ[39m The deprecated feature was likely used in the [34mSeurat[39m package.
  Please report the issue at [3m[34m&lt;https://github.com/satijalab/seurat/issues&gt;[39m[23m.‚Äù
Warning message:
‚Äú[1m[22mThe `slot` argument of `SetAssayData()` is deprecated as of SeuratObject 5.0.0.
[36m‚Ñπ[39m Please use the `layer` argument instead.
[36m‚Ñπ[39m The deprecated feature was likely used in the [34mSeurat[39m package.
  Please report the issue at [3m[34m&lt;https://github.com/satijalab/seurat/issues&gt;[39m[23m.‚Äù
</code></pre>
<p>Once integrative analysis is complete, you can rejoin the layers - which collapses the individual datasets together and recreates the original <code>counts</code> and <code>data</code> layers. You will need to do this before performing any differential expression analysis. However, you can always resplit the layers in case you would like to reperform integrative analysis.</p>
<pre><code class="language-R"># re-join layers after integration
so_merged_integ[[&quot;RNA&quot;]] &lt;- JoinLayers(so_merged_integ[[&quot;RNA&quot;]])

</code></pre>
<pre><code class="language-R">so_merged_integ &lt;- FindNeighbors(so_merged_integ,
                         reduction = &quot;integrated.cca&quot;, dims = 1:30)
so_merged_integ &lt;- FindClusters(so_merged_integ, resolution = 2)
</code></pre>
<pre><code>Computing nearest neighbor graph

Computing SNN



Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 6000
Number of edges: 203245

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8375
Number of communities: 37
Elapsed time: 0 seconds
</code></pre>
<pre><code class="language-R">so_merged_integ &lt;- RunUMAP(so_merged_integ,
                          dims = 1:30,
                          reduction = &quot;integrated.cca&quot;)

</code></pre>
<pre><code>11:41:23 UMAP embedding parameters a = 0.9922 b = 1.112

11:41:23 Read 6000 rows and found 30 numeric columns

11:41:23 Using Annoy for neighbor search, n_neighbors = 30

11:41:23 Building Annoy index with metric = cosine, n_trees = 50

0%   10   20   30   40   50   60   70   80   90   100%

[----|----|----|----|----|----|----|----|----|----|

*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
|

11:41:24 Writing NN index file to temp file /tmp/RtmpCzdl7E/file783386fd3f3

11:41:24 Searching Annoy index using 1 thread, search_k = 3000

11:41:27 Annoy recall = 100%

11:41:28 Commencing smooth kNN distance calibration using 1 thread
 with target n_neighbors = 30

11:41:31 Initializing from normalized Laplacian + noise (using RSpectra)

11:41:31 Commencing optimization for 500 epochs, with 243902 positive edges

11:41:31 Using rng type: pcg

11:41:41 Optimization finished
</code></pre>
<pre><code class="language-R">DimPlot(so_merged_integ,
             reduction = &quot;umap&quot;,
             split.by = c(&quot;orig.ident&quot;))
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_cele_files/BIOI611_scRNA_cele_77_0.png" /></p>
<pre><code class="language-R">DefaultAssay(so_merged_integ)
</code></pre>
<p>'RNA'</p>
<pre><code class="language-R">markers &lt;- FindAllMarkers(so_merged_integ)
</code></pre>
<pre><code>Calculating cluster 0

Calculating cluster 1

Calculating cluster 2

Calculating cluster 3

Calculating cluster 4

Calculating cluster 5

Calculating cluster 6

Calculating cluster 7

Calculating cluster 8

Calculating cluster 9

Calculating cluster 10

Calculating cluster 11

Calculating cluster 12

Calculating cluster 13

Calculating cluster 14

Calculating cluster 15

Calculating cluster 16

Calculating cluster 17

Calculating cluster 18

Calculating cluster 19

Calculating cluster 20

Calculating cluster 21

Calculating cluster 22

Calculating cluster 23

Calculating cluster 24

Calculating cluster 25

Calculating cluster 26

Calculating cluster 27

Calculating cluster 28

Calculating cluster 29

Calculating cluster 30

Calculating cluster 31

Calculating cluster 32

Calculating cluster 33

Calculating cluster 34

Calculating cluster 35

Calculating cluster 36
</code></pre>
<pre><code class="language-R">markers %&gt;%
    group_by(cluster) %&gt;%
    dplyr::filter(avg_log2FC &gt; 1) %&gt;%
    slice_head(n = 2) %&gt;%
    ungroup() -&gt; top2
DoHeatmap(so_merged_integ, features = top2$gene) + NoLegend()
</code></pre>
<pre><code>Warning message in DoHeatmap(so_merged_integ, features = top2$gene):
‚ÄúThe following features were omitted as they were not found in the scale.data slot for the RNA assay: nmur-3, timp-1, acp-6, C14A4.6, mltn-13, srw-12, T03G11.9, C02F5.2, C33D9.10, lev-9, C08F1.6, cank-26‚Äù
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_cele_files/BIOI611_scRNA_cele_80_1.png" /></p>
<pre><code class="language-R">head(markers)
</code></pre>
<table class="dataframe">
<caption>A data.frame: 6 √ó 7</caption>
<thead>
    <tr><th></th><th scope=col>p_val</th><th scope=col>avg_log2FC</th><th scope=col>pct.1</th><th scope=col>pct.2</th><th scope=col>p_val_adj</th><th scope=col>cluster</th><th scope=col>gene</th></tr>
    <tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
    <tr><th scope=row>asp-4</th><td>1.528372e-113</td><td>1.650373</td><td>0.942</td><td>0.461</td><td>3.054452e-109</td><td>0</td><td>asp-4   </td></tr>
    <tr><th scope=row>ZK930.6</th><td>2.165015e-107</td><td>2.966635</td><td>0.540</td><td>0.133</td><td>4.326783e-103</td><td>0</td><td>ZK930.6 </td></tr>
    <tr><th scope=row>ins-2</th><td> 1.469696e-91</td><td>4.274237</td><td>0.462</td><td>0.114</td><td> 2.937187e-87</td><td>0</td><td>ins-2   </td></tr>
    <tr><th scope=row>hil-2</th><td> 2.536414e-79</td><td>1.401867</td><td>0.964</td><td>0.766</td><td> 5.069024e-75</td><td>0</td><td>hil-2   </td></tr>
    <tr><th scope=row>C08F1.10</th><td> 3.235846e-79</td><td>1.359935</td><td>0.916</td><td>0.537</td><td> 6.466837e-75</td><td>0</td><td>C08F1.10</td></tr>
    <tr><th scope=row>hil-3</th><td> 3.234851e-78</td><td>1.962521</td><td>0.830</td><td>0.447</td><td> 6.464851e-74</td><td>0</td><td>hil-3   </td></tr>
</tbody>
</table>

<pre><code class="language-R">FeaturePlot(so_merged_integ, reduction = &quot;umap&quot;, feature=&quot;ham-1&quot;, pt.size = 0.1,
             split.by = c(&quot;orig.ident&quot;))
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_cele_files/BIOI611_scRNA_cele_82_0.png" /></p>
<pre><code class="language-R">so_merged_integ
</code></pre>
<pre><code>An object of class Seurat 
19985 features across 6000 samples within 1 assay 
Active assay: RNA (19985 features, 2000 variable features)
 3 layers present: data, counts, scale.data
 4 dimensional reductions calculated: pca, umap.unintegrated, integrated.cca, umap
</code></pre>
<pre><code class="language-R">FeaturePlot(so_merged_integ, reduction = &quot;umap&quot;, feature=&quot;egl-21&quot;, pt.size = 0.1,
             split.by = c(&quot;orig.ident&quot;))
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_cele_files/BIOI611_scRNA_cele_84_0.png" /></p>
<pre><code class="language-R">FeaturePlot(so_merged_integ, reduction = &quot;umap&quot;, feature=&quot;dsl-3&quot;, pt.size = 0.1,
             split.by = c(&quot;orig.ident&quot;))
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_cele_files/BIOI611_scRNA_cele_85_0.png" /></p>
<p>With the integrated Seurat object, you can perform cell type annotation using marker genes.</p>
<p>‚ö†Ô∏è Important</p>
<p>In this class, we will not perform cell type annotation for this dataset. With the notebook for analyzing the human PBMC data, you will be able to perform cell type annnotation if needed.</p>
<p>In the next section, you will learn how to perform trajectory and pseudotime analysis. You will directly utilize the annotation performed by the authors who generated this dataset in <a href="https://www.science.org/doi/10.1126/science.aax1971">the Science paper</a>.</p>
<p>The data can be obtained using the URLs below:</p>
<pre><code>## count matrix
https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_expression.rds
## metadata
https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_colData.rds
## gene list
https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_rowData.rds
</code></pre>
<h2 id="save-the-seurat-object">Save the Seurat object</h2>
<pre><code class="language-R">saveRDS(so_merged_integ, file = &quot;Seurat_object_10x_cele_final.rds&quot;)
</code></pre>
<pre><code>Found more than one class "package_version" in cache; using the first, from namespace 'SeuratObject'

Also defined by ‚Äòalabaster.base‚Äô

Found more than one class "package_version" in cache; using the first, from namespace 'SeuratObject'

Also defined by ‚Äòalabaster.base‚Äô
</code></pre>
<pre><code class="language-R">remotes::install_github(&quot;10xGenomics/loupeR&quot;)
loupeR::setup()
</code></pre>
<pre><code>Skipping install of 'loupeR' from a github remote, the SHA1 (d7994c5c) has not changed since last install.
  Use `force = TRUE` to force installation

Warning message in fun(libname, pkgname):
‚ÄúPlease call `loupeR::setup()` to install the Louper executable  and to agree to the EULA before continuing‚Äù



Installing Executable



2025/12/04 11:43:04 Downloading executable



The LoupeR executable is subject to the 10x End User Software License, available at:
https://10xgen.com/EULA

Do you accept the End-User License Agreement
(y/yes or n/no): yes

EULA
</code></pre>
<pre><code class="language-R">library(loupeR)
</code></pre>
<pre><code class="language-R">create_loupe_from_seurat(so_merged_integ,
                output_name = &quot;Seurat_object_10x_cele_merged_integ&quot;,
                force = TRUE)

</code></pre>
<pre><code>2025/12/04 11:53:23 extracting matrix, clusters, and projections

2025/12/04 11:53:24 selected assay: RNA

2025/12/04 11:53:24 selected clusters: active_cluster orig.ident unintegrated_clusters seurat_clusters RNA_snn_res.2

2025/12/04 11:53:24 selected projections: umap.unintegrated umap

2025/12/04 11:53:24 validating count matrix

2025/12/04 11:53:24 validating clusters

2025/12/04 11:53:24 validating projections

2025/12/04 11:53:24 creating temporary hdf5 file: /tmp/RtmpCzdl7E/file7833a04d0eb.h5

2025/12/04 11:53:27 invoking louper executable

2025/12/04 11:53:27 running command: "/root/.local/share/R/loupeR/louper create --input='/tmp/RtmpCzdl7E/file7833a04d0eb.h5' --output='Seurat_object_10x_cele_merged_integ.cloupe' --force"
</code></pre>
<h2 id="reference">Reference</h2>
<p>https://monashbioinformaticsplatform.github.io/Single-Cell-Workshop/pbmc3k_tutorial.html</p>
<p>https://bioinformatics.ccr.cancer.gov/docs/getting-started-with-scrna-seq/IntroToR_Seurat/</p>
<p>https://hbctraining.github.io/scRNA-seq/lessons/elbow_plot_metric.html</p>
<p>https://satijalab.org/seurat/articles/integration_introduction</p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../BIOI611_scRNA_seq_cele_cellranger/" class="btn btn-neutral float-left" title="Initial analysis of 10x scRNA-seq data for C. elegans using cellranger"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../bioi611_monocle_cele/" class="btn btn-neutral float-right" title="Downstream analysis - trajectory analysis using Monocle3">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>Copyright &copy; 2025 Shaojun Xie</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/Bix4UMD/BIOI611_lab.git" class="fa fa-code-fork" style="color: #fcfcfc"> Bix4UMD/BIOI611_lab</a>
        </span>
    
    
      <span><a href="../BIOI611_scRNA_seq_cele_cellranger/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../bioi611_monocle_cele/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../javascripts/mathjax.js"></script>
      <script src="https://unpkg.com/mathjax@3/es5/tex-mml-chtml.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
