<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Shaojun Xie" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>BIOI611 scRNA 2024 - Lab note for UMD BIOI611</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "BIOI611 scRNA 2024";
        var mkdocs_page_input_path = "BIOI611_scRNA_2024.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/yaml.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/rust.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> Lab note for UMD BIOI611
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption"><span class="caption-text">Get ready for the course</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../basic_linux/">Basic Linux</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Bulk RNA-seq</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../bulkRNAseq_lab/">Read QC and alignment</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Phred_FQ/">Phred score in FASTQ</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_introR/">Minimal R Introduction</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_DESeq2_analysis/">DEG analysis using DESeq2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_DAVID/">Pathway analysis using DAVID</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_ballgown_example/">DE analysis at isoform-leve using ballgown (example data)</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_bulkRNA_SE_ballgown/">DE analysis at isoform-leve using ballgown (C. ele data)</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_long_read_transcriptome/">Long read transcriptome</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">scRNA-seq</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_analysis_10x_dataset_PBMC/">Initial analysis of 10x scRNA-seq data for human PBMC using cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_scRNA/">Downstream analysis of 10x scRNA-seq data for human PBMC using Seurat</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_scRNA_seq_cele_cellranger/">Initial analysis of 10x scRNA-seq data for C. elegans using cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_scRNA_cele/">Downstream analysis of 10x scRNA-seq data for C. elegans data</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bioi611_monocle_cele/">Downstream analysis - trajectory analysis using Monocle3</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">Lab note for UMD BIOI611</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">BIOI611 scRNA 2024</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/Bix4UMD/BIOI611_lab.git/edit/master/docs/BIOI611_scRNA_2024.md">Edit on Bix4UMD/BIOI611_lab</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <p><a href="https://colab.research.google.com/github/Bix4UMD/BIOI611_lab/blob/main/docs/BIOI611_scRNA_2024.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<h1 id="_1"></h1>
<h2 id="install-required-r-packages">Install required R packages</h2>
<pre><code class="language-R"># # Install the remotes package
# if (!requireNamespace(&quot;remotes&quot;, quietly = TRUE)) {
#   install.packages(&quot;remotes&quot;)
# }
# # Install Seurat
# if (!requireNamespace(&quot;Seurat&quot;, quietly = TRUE)) {
#     remotes::install_github(&quot;satijalab/seurat&quot;, &quot;seurat5&quot;, quiet = TRUE)
# }
# # Install BiocManager
# if (!require(&quot;BiocManager&quot;, quietly = TRUE))
#     install.packages(&quot;BiocManager&quot;)

# # Install SingleR package
# if (!require(&quot;hdf5r&quot;, quietly = TRUE)){
#     BiocManager::install(&quot;hdf5r&quot;)
# }
# # Install SingleR package
# if (!require(&quot;presto&quot;, quietly = TRUE)){
#     remotes::install_github(&quot;immunogenomics/presto&quot;)
# }
# # Install SingleR package
# if (!require(&quot;SingleR&quot;, quietly = TRUE)){
#     BiocManager::install(&quot;SingleR&quot;)
# }
# if (!require(&quot;celldex&quot;, quietly = TRUE)){
#     BiocManager::install(&quot;celldex&quot;)
# }
# if (!require(&quot;SingleCellExperiment&quot;, quietly = TRUE)){
#     BiocManager::install(&quot;SingleCellExperiment&quot;)
# }
# if (!require(&quot;scater&quot;, quietly = TRUE)){
#     BiocManager::install(&quot;scater&quot;)
# }
</code></pre>
<pre><code class="language-R">## Installing the R packages could take around 51 minutes
## To speed up this process, you can download the R lib files
## saved from a working Google Colab session
## https://drive.google.com/file/d/1EQvZnsV6P0eNjbW0hwYhz0P5z0iH3bsL/view?usp=drive_link
system(&quot;gdown 1EQvZnsV6P0eNjbW0hwYhz0P5z0iH3bsL&quot;, intern = TRUE)
</code></pre>
<pre><code class="language-R">system(&quot;md5sum R_lib4scRNA.tar.gz&quot;, intern = TRUE)
</code></pre>
<p><span style=white-space:pre-wrap>'5898c04fca5e680710cd6728ef9b1422  R_lib4scRNA.tar.gz'</span></p>
<pre><code class="language-R">## required by scater package
system(&quot;apt-get install libx11-dev libcairo2-dev&quot;, intern = TRUE)
system(&quot;tar zxvf R_lib4scRNA.tar.gz&quot;)

</code></pre>
<pre><code class="language-R">## required by scater package
system(&quot;apt-get install libx11-dev libcairo2-dev&quot;, intern = TRUE)
system(&quot;tar zxvf BIOI611_custom_R_lib_Nov_2025.tar.gz&quot;)
</code></pre>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'Reading package lists...'</li><li>'Building dependency tree...'</li><li>'Reading state information...'</li><li>'libcairo2-dev is already the newest version (1.16.0-5ubuntu2).'</li><li>'libx11-dev is already the newest version (2:1.7.5-1ubuntu0.3).'</li><li>'0 upgraded, 0 newly installed, 0 to remove and 41 not upgraded.'</li></ol>

<pre><code class="language-R">
</code></pre>
<pre><code class="language-R">.libPaths(c(&quot;/content/usr/local/lib/R/site-library&quot;, .libPaths()))
.libPaths()
</code></pre>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'/content/usr/local/lib/R/site-library'</li><li>'/usr/local/lib/R/site-library'</li><li>'/usr/lib/R/site-library'</li><li>'/usr/lib/R/library'</li></ol>

<h2 id="load-required-r-packages">Load required R packages</h2>
<pre><code class="language-R">library(Seurat)
library(dplyr)
library(SingleR)
library(celldex)
library(scater)
library(SingleCellExperiment)
</code></pre>
<pre><code class="language-R">list.files()
</code></pre>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'BIOI611_custom_R_lib_Nov_2025.tar.gz'</li><li>'filtered_feature_bc_matrix.h5'</li><li>'sample_data'</li><li>'sessionInfo.txt'</li><li>'usr'</li></ol>

<pre><code class="language-R"># https://drive.google.com/file/d/1-CvmcLvKMYW-OcLuGfFuGQMK2b5_VMFk/view?usp=drive_link
# Download &quot;filtered_feature_bc_matrix.h5&quot;
# Output of cellranger
system(&quot;gdown 1-CvmcLvKMYW-OcLuGfFuGQMK2b5_VMFk&quot;, intern = TRUE)
</code></pre>
<pre><code class="language-R">system(&quot;md5sum filtered_feature_bc_matrix.h5&quot;, intern = TRUE)
</code></pre>
<p><span style=white-space:pre-wrap>'360fc0760ebb9e6dd253d808a427b20d  filtered_feature_bc_matrix.h5'</span></p>
<pre><code class="language-R">count_mtx_scrna &lt;- Read10X_h5(&quot;filtered_feature_bc_matrix.h5&quot;)
# If you have the filtered_feature_bc_matrix/ folder, you can use
# Read10X to create 'count_mtx_scrna'
# system(&quot;mkdir filtered_feature_bc_matrix/; mv filtered_feature_bc_matrix.zip filtered_feature_bc_matrix&quot;)
# system(&quot;cd filtered_feature_bc_matrix; unzip filtered_feature_bc_matrix.zip&quot;)
# count_mtx_scrna &lt;- Read10X(&quot;filtered_feature_bc_matrix/&quot;)
</code></pre>
<pre><code class="language-R">class(count_mtx_scrna)
</code></pre>
<p>'dgCMatrix'</p>
<p>The dgCMatrix class is a specific data structure in R's Matrix package, designed to store sparse matrices in a memory-efficient format. Sparse matrices are those with many zeros, making them ideal for high-dimensional data in applications like bioinformatics, where gene expression matrices often contain a lot of zeroes.</p>
<blockquote>
<p>Why Use dgCMatrix?</p>
</blockquote>
<ul>
<li><strong>Memory Efficiency</strong>: Storing only non-zero values saves memory, especially in high-dimensional matrices.</li>
<li><strong>Computational Speed</strong>: Some operations on sparse matrices can be faster, as computations are limited to non-zero entries.</li>
</ul>
<pre><code class="language-R">print(format(object.size(count_mtx_scrna), units = &quot;MB&quot;))
</code></pre>
<pre><code>[1] "168.7 Mb"
</code></pre>
<pre><code class="language-R"># Check a few genes in the first 20 cells
count_mtx_scrna[c(&quot;CD3D&quot;, &quot;TCL1A&quot;, &quot;MS4A1&quot;), 100:140]
</code></pre>
<pre><code>  [[ suppressing 41 column names ‘AACCATGCACTCAAGT-1’, ‘AACCATGGTAGCTTGT-1’, ‘AACCATGTCAATCCGA-1’ ... ]]




3 x 41 sparse Matrix of class "dgCMatrix"

CD3D  8 1 . . . 3 . 2 10  . 3 . . . 2 7 1 . . . 1 1  . 8 . . 2 4 . . . 12 11 .
TCL1A . . . . . . 6 .  .  . . . . . . . . . . . . . 10 . . . . . . . .  .  . .
MS4A1 . . . . . . 9 .  . 16 . . . . . . . . . . . .  5 . . . . . . . .  .  . .

CD3D   . 5 . .  . 3 .
TCL1A  . . . .  . . .
MS4A1 20 . . . 39 . .
</code></pre>
<pre><code class="language-R"># non-normalized da# Initialize the Seurat object with the raw count matrix
pbmc &lt;- CreateSeuratObject(counts = count_mtx_scrna,
                        project = &quot;pbmc5k&quot;,
                        min.cells = 3, min.features = 200)
pbmc
</code></pre>
<pre><code>An object of class Seurat 
24785 features across 4884 samples within 1 assay 
Active assay: RNA (24785 features, 0 variable features)
 1 layer present: counts
</code></pre>
<h2 id="understand-seurat-object">Understand Seurat object</h2>
<blockquote>
<p>Seurat slots
https://github.com/satijalab/seurat/wiki/seurat</p>
</blockquote>
<pre><code class="language-R">str(pbmc)
</code></pre>
<pre><code>Formal class 'Seurat' [package "SeuratObject"] with 13 slots
  ..@ assays      :List of 1
  .. ..$ RNA:Formal class 'Assay5' [package "SeuratObject"] with 8 slots
  .. .. .. ..@ layers    :List of 1
  .. .. .. .. ..$ counts:Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
  .. .. .. .. .. .. ..@ i       : int [1:14449622] 6 17 42 62 79 83 85 94 100 109 ...
  .. .. .. .. .. .. ..@ p       : int [1:4885] 0 3378 5344 5581 8581 10897 13921 16902 20299 22669 ...
  .. .. .. .. .. .. ..@ Dim     : int [1:2] 24785 4884
  .. .. .. .. .. .. ..@ Dimnames:List of 2
  .. .. .. .. .. .. .. ..$ : NULL
  .. .. .. .. .. .. .. ..$ : NULL
  .. .. .. .. .. .. ..@ x       : num [1:14449622] 1 1 4 1 1 2 1 1 1 1 ...
  .. .. .. .. .. .. ..@ factors : list()
  .. .. .. ..@ cells     :Formal class 'LogMap' [package "SeuratObject"] with 1 slot
  .. .. .. .. .. ..@ .Data: logi [1:4884, 1] TRUE TRUE TRUE TRUE TRUE TRUE ...
  .. .. .. .. .. .. ..- attr(*, "dimnames")=List of 2
  .. .. .. .. .. .. .. ..$ : chr [1:4884] "AAACCCATCAGATGCT-1" "AAACGAAAGTGCTACT-1" "AAACGAAGTCGTAATC-1" "AAACGAAGTTGCCAAT-1" ...
  .. .. .. .. .. .. .. ..$ : chr "counts"
  .. .. .. .. .. ..$ dim     : int [1:2] 4884 1
  .. .. .. .. .. ..$ dimnames:List of 2
  .. .. .. .. .. .. ..$ : chr [1:4884] "AAACCCATCAGATGCT-1" "AAACGAAAGTGCTACT-1" "AAACGAAGTCGTAATC-1" "AAACGAAGTTGCCAAT-1" ...
  .. .. .. .. .. .. ..$ : chr "counts"
  .. .. .. ..@ features  :Formal class 'LogMap' [package "SeuratObject"] with 1 slot
  .. .. .. .. .. ..@ .Data: logi [1:24785, 1] TRUE TRUE TRUE TRUE TRUE TRUE ...
  .. .. .. .. .. .. ..- attr(*, "dimnames")=List of 2
  .. .. .. .. .. .. .. ..$ : chr [1:24785] "ENSG00000238009" "ENSG00000241860" "ENSG00000290385" "ENSG00000291215" ...
  .. .. .. .. .. .. .. ..$ : chr "counts"
  .. .. .. .. .. ..$ dim     : int [1:2] 24785 1
  .. .. .. .. .. ..$ dimnames:List of 2
  .. .. .. .. .. .. ..$ : chr [1:24785] "ENSG00000238009" "ENSG00000241860" "ENSG00000290385" "ENSG00000291215" ...
  .. .. .. .. .. .. ..$ : chr "counts"
  .. .. .. ..@ default   : int 1
  .. .. .. ..@ assay.orig: chr(0) 
  .. .. .. ..@ meta.data :'data.frame': 24785 obs. of  0 variables
  .. .. .. ..@ misc      :List of 1
  .. .. .. .. ..$ calcN: logi TRUE
  .. .. .. ..@ key       : chr "rna_"
  ..@ meta.data   :'data.frame':    4884 obs. of  3 variables:
  .. ..$ orig.ident  : Factor w/ 1 level "pbmc5k": 1 1 1 1 1 1 1 1 1 1 ...
  .. ..$ nCount_RNA  : num [1:4884] 11578 5655 14728 10903 6174 ...
  .. ..$ nFeature_RNA: int [1:4884] 3378 1966 237 3000 2316 3024 2981 3397 2370 2811 ...
  ..@ active.assay: chr "RNA"
  ..@ active.ident: Factor w/ 1 level "pbmc5k": 1 1 1 1 1 1 1 1 1 1 ...
  .. ..- attr(*, "names")= chr [1:4884] "AAACCCATCAGATGCT-1" "AAACGAAAGTGCTACT-1" "AAACGAAGTCGTAATC-1" "AAACGAAGTTGCCAAT-1" ...
  ..@ graphs      : list()
  ..@ neighbors   : list()
  ..@ reductions  : list()
  ..@ images      : list()
  ..@ project.name: chr "pbmc5k"
  ..@ misc        : list()
  ..@ version     :Classes 'package_version', 'numeric_version'  hidden list of 1
  .. ..$ : int [1:3] 5 2 0
  ..@ commands    : list()
  ..@ tools       : list()
</code></pre>
<pre><code class="language-R">
</code></pre>
<pre><code class="language-R">slotNames(pbmc)
</code></pre>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'assays'</li><li>'meta.data'</li><li>'active.assay'</li><li>'active.ident'</li><li>'graphs'</li><li>'neighbors'</li><li>'reductions'</li><li>'images'</li><li>'project.name'</li><li>'misc'</li><li>'version'</li><li>'commands'</li><li>'tools'</li></ol>

<pre><code class="language-R">
</code></pre>
<h3 id="access-seurat-object">Access Seurat object</h3>
<pre><code class="language-R">pbmc@active.assay
</code></pre>
<p>'RNA'</p>
<pre><code class="language-R">class(pbmc@meta.data)
head(pbmc@meta.data, 4)
</code></pre>
<p>'data.frame'</p>
<table class="dataframe">
<caption>A data.frame: 4 × 3</caption>
<thead>
    <tr><th></th><th scope=col>orig.ident</th><th scope=col>nCount_RNA</th><th scope=col>nFeature_RNA</th></tr>
    <tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>
</thead>
<tbody>
    <tr><th scope=row>AAACCCATCAGATGCT-1</th><td>pbmc5k</td><td>11578</td><td>3378</td></tr>
    <tr><th scope=row>AAACGAAAGTGCTACT-1</th><td>pbmc5k</td><td> 5655</td><td>1966</td></tr>
    <tr><th scope=row>AAACGAAGTCGTAATC-1</th><td>pbmc5k</td><td>14728</td><td> 237</td></tr>
    <tr><th scope=row>AAACGAAGTTGCCAAT-1</th><td>pbmc5k</td><td>10903</td><td>3000</td></tr>
</tbody>
</table>

<pre><code class="language-R">colSums(pbmc@assays$RNA$counts)[1:3]
</code></pre>
<style>
.dl-inline {width: auto; margin:0; padding: 0}
.dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}
.dl-inline>dt::after {content: ":\0020"; padding-right: .5ex}
.dl-inline>dt:not(:first-of-type) {padding-left: .5ex}
</style>
<dl class=dl-inline><dt>AAACCCATCAGATGCT-1</dt><dd>11578</dd><dt>AAACGAAAGTGCTACT-1</dt><dd>5655</dd><dt>AAACGAAGTCGTAATC-1</dt><dd>14728</dd></dl>

<pre><code class="language-R">Layers(pbmc)
</code></pre>
<p>'counts'</p>
<h2 id="data-preprocessing">Data preprocessing</h2>
<pre><code class="language-R"># Use $ operator to add columns to object metadata.
pbmc$percent.mt &lt;- PercentageFeatureSet(pbmc,
                           pattern = &quot;^MT-&quot;)

</code></pre>
<pre><code class="language-R">colnames(pbmc@meta.data)
</code></pre>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'orig.ident'</li><li>'nCount_RNA'</li><li>'nFeature_RNA'</li><li>'percent.mt'</li></ol>

<pre><code class="language-R">head(pbmc, 4)
</code></pre>
<table class="dataframe">
<caption>A data.frame: 4 × 4</caption>
<thead>
    <tr><th></th><th scope=col>orig.ident</th><th scope=col>nCount_RNA</th><th scope=col>nFeature_RNA</th><th scope=col>percent.mt</th></tr>
    <tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
    <tr><th scope=row>AAACCCATCAGATGCT-1</th><td>pbmc5k</td><td>11578</td><td>3378</td><td>3.62756953</td></tr>
    <tr><th scope=row>AAACGAAAGTGCTACT-1</th><td>pbmc5k</td><td> 5655</td><td>1966</td><td>4.29708223</td></tr>
    <tr><th scope=row>AAACGAAGTCGTAATC-1</th><td>pbmc5k</td><td>14728</td><td> 237</td><td>0.02715915</td></tr>
    <tr><th scope=row>AAACGAAGTTGCCAAT-1</th><td>pbmc5k</td><td>10903</td><td>3000</td><td>4.89773457</td></tr>
</tbody>
</table>

<pre><code class="language-R">pbmc
</code></pre>
<pre><code>An object of class Seurat 
24785 features across 4884 samples within 1 assay 
Active assay: RNA (24785 features, 0 variable features)
 1 layer present: counts
</code></pre>
<pre><code class="language-R">NewVlnPlot &lt;- function(seurat_obj, col = 3){
  # Use violin plot to visualize QC metrics
  tem_plots &lt;- VlnPlot(
    seurat_obj,
    features = c(&quot;nFeature_RNA&quot;, &quot;nCount_RNA&quot;, &quot;percent.mt&quot;),
    combine = FALSE  # return list of ggplot objects
  )
  # Apply theme to each plot
  tem_plots &lt;- lapply(tem_plots, function(x) x + NoLegend())   # OR + theme_bw()
  # Combine again
  patchwork::wrap_plots(tem_plots, ncol = 3)
}
</code></pre>
<pre><code class="language-R"># VlnPlot(
#     pbmc,
#     features = c(&quot;nFeature_RNA&quot;, &quot;nCount_RNA&quot;, &quot;percent.mt&quot;),
#     col = 3
#   )

NewVlnPlot(pbmc, col = 3)
</code></pre>
<pre><code>Warning message:
“Default search for "data" layer in "RNA" assay yielded no results; utilizing "counts" layer instead.”
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_2024_files/BIOI611_scRNA_2024_36_1.png" /></p>
<h3 id="how-to-read-the-violin-plot">How to read the Violin Plot</h3>
<ul>
<li>
<p>Shape: Each violin plot shows the distribution of values for each feature across the cells in your dataset. The shape of the plot indicates the density of cells with particular values for that feature.</p>
<p>Wider sections indicate more cells with those values.
Narrow sections indicate fewer cells with those values.</p>
</li>
<li>
<p>Vertical Axis: Represents the range of values for each feature. For instance:</p>
</li>
</ul>
<p><code>nFeature_RNA</code> and <code>nCount_RNA</code>: Higher values suggest more gene diversity and RNA content, respectively.</p>
<p><code>percent.mt</code>: Higher values indicate higher mitochondrial content, which may point to stressed or dying cells.</p>
<ul>
<li>Horizontal Axis (Groups): If your dataset is separated into clusters or groups (e.g., cell types or conditions), each group will have its own violin, allowing you to compare distributions between groups.</li>
</ul>
<h3 id="how-to-interpret-qc-plot">How to interpret QC plot</h3>
<p><code>nFeature_RNA</code>: The number of unique features (genes) detected per cell.</p>
<p>Extremely high values could suggest potential doublets (two cells mistakenly captured as one), as two cells would have more unique genes combined.</p>
<p>Low number of detected genes - potential ambient mRNA (not real cells)</p>
<p><code>nCount_RNA</code>: The total number of RNA molecules (or unique molecular identifiers, UMIs) detected per cell.</p>
<p>Higher counts generally indicate higher RNA content, but they could also result from cell doublets.
Cells with very low nCount_RNA might represent poor-quality cells with low RNA capture, while very high counts may also suggest doublets.</p>
<p><code>percent.mt</code>: The percentage of reads mapping to mitochondrial genes.</p>
<p>High mitochondrial content often indicates cell stress or apoptosis, as damaged cells tend to release mitochondrial RNA.</p>
<p>Filtering cells with high <code>percent.mt</code> values is common to exclude potentially dying cells.</p>
<pre><code class="language-R"># FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 &lt;- FeatureScatter(pbmc, feature1 = &quot;nCount_RNA&quot;, feature2 = &quot;percent.mt&quot;)
plot2 &lt;- FeatureScatter(pbmc, feature1 = &quot;nCount_RNA&quot;, feature2 = &quot;nFeature_RNA&quot;)
plot1 + plot2
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_2024_files/BIOI611_scRNA_2024_39_0.png" /></p>
<pre><code class="language-R"># Load necessary libraries
library(Seurat)
library(ggplot2)

# Define the function to calculate median and MAD values
calculate_thresholds &lt;- function(seurat_obj) {
  # Extract relevant columns
  nFeature_values &lt;- seurat_obj@meta.data$nFeature_RNA
  nCount_values &lt;- seurat_obj@meta.data$nCount_RNA
  percent_mt_values &lt;- seurat_obj@meta.data$percent.mt

  # Calculate medians and MADs
  nFeature_median &lt;- median(nFeature_values, na.rm = TRUE)
  nFeature_mad &lt;- mad(nFeature_values, constant = 1, na.rm = TRUE)

  nCount_median &lt;- median(nCount_values, na.rm = TRUE)
  nCount_mad &lt;- mad(nCount_values, constant = 1, na.rm = TRUE)

  percent_mt_median &lt;- median(percent_mt_values, na.rm = TRUE)
  percent_mt_mad &lt;- mad(percent_mt_values, constant = 1, na.rm = TRUE)

  # Calculate thresholds for horizontal lines
  thresholds &lt;- list(
    nFeature_upper = nFeature_median + 4 * nFeature_mad,
    nFeature_lower = nFeature_median - 4 * nFeature_mad,
    nCount_upper = nCount_median + 4 * nCount_mad,
    nCount_lower = nCount_median - 4 * nCount_mad,
    percent_mt_upper = percent_mt_median + 4 * percent_mt_mad
  )

  return(thresholds)
}

# Calculate thresholds
thresholds &lt;- calculate_thresholds(pbmc)

</code></pre>
<pre><code class="language-R">thresholds
</code></pre>
<dl>
    <dt>$nFeature_upper</dt>
        <dd>5243.5</dd>
    <dt>$nFeature_lower</dt>
        <dd>583.5</dd>
    <dt>$nCount_upper</dt>
        <dd>19044</dd>
    <dt>$nCount_lower</dt>
        <dd>-1224</dd>
    <dt>$percent_mt_upper</dt>
        <dd>8.44783722411371</dd>
</dl>

<pre><code class="language-R">vplot1 &lt;- VlnPlot(pbmc, features = c(&quot;nFeature_RNA&quot;), ncol = 2) +
   geom_hline(yintercept = thresholds$nFeature_upper,
              color = &quot;blue&quot;, linetype = &quot;solid&quot;) +
   geom_hline(yintercept = thresholds$nFeature_lower,
              color = &quot;blue&quot;, linetype = &quot;solid&quot;) +
              theme(legend.position=&quot;none&quot;)
vplot2 &lt;- VlnPlot(pbmc, features = c(&quot;percent.mt&quot;), ncol = 2) +
   geom_hline(yintercept = thresholds$percent_mt_upper,
              color = &quot;blue&quot;, linetype = &quot;solid&quot;) +
              theme(legend.position=&quot;none&quot;)
vplot1 + vplot2
</code></pre>
<pre><code>Warning message:
“Default search for "data" layer in "RNA" assay yielded no results; utilizing "counts" layer instead.”
Warning message:
“Default search for "data" layer in "RNA" assay yielded no results; utilizing "counts" layer instead.”
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_2024_files/BIOI611_scRNA_2024_42_1.png" /></p>
<h3 id="filter-out-potential-doublets-empty-droplets-and-dying-cells">Filter out potential doublets, empty droplets and dying cells</h3>
<pre><code class="language-R">pbmc &lt;- subset(pbmc,
                subset =  nFeature_RNA &gt; thresholds$nCount_lower &amp;
                nFeature_RNA &lt; thresholds$nFeature_upper &amp;
                percent.mt &lt; thresholds$percent_mt_upper)

</code></pre>
<pre><code class="language-R"># Use violin plot to visualize QC metrics after QC
# VlnPlot(pbmc,
#         features = c(&quot;nFeature_RNA&quot;, &quot;nCount_RNA&quot;, &quot;percent.mt&quot;),
#         ncol = 3)
NewVlnPlot(pbmc, col = 3)
</code></pre>
<pre><code>Warning message:
“Default search for "data" layer in "RNA" assay yielded no results; utilizing "counts" layer instead.”
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_2024_files/BIOI611_scRNA_2024_45_1.png" /></p>
<p>Instead of using an arbitrary number, you can also use statistical algorithm to predict doublets and empty droplets to filter the cells, such as <code>DoubletFinder</code> and <code>EmptyDrops</code>.</p>
<h3 id="normalization-and-scaling-of-the-data">Normalization and Scaling of the data</h3>
<h3 id="normalization">Normalization</h3>
<p>After removing unwanted cells from the dataset, the next step is to normalize the data. By default, a global-scaling normalization method “LogNormalize” that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. In Seurat v5, Normalized values are stored in <code>pbmc[["RNA"]]$data</code>.</p>
<pre><code class="language-R">pbmc &lt;- NormalizeData(pbmc) # normalization.method = &quot;LogNormalize&quot;, scale.factor = 10000
</code></pre>
<pre><code>Normalizing layer: counts
</code></pre>
<p>While this method of normalization is standard and widely used in scRNA-seq analysis, global-scaling relies on an assumption that each cell originally contains the same number of RNA molecules.</p>
<p>Next, we identify a subset of features that show high variation across cells in the dataset—meaning they are highly expressed in some cells and lowly expressed in others. Prior work, including our own, has shown that focusing on these variable genes in downstream analyses can enhance the detection of biological signals in single-cell datasets.</p>
<p>The approach used in Seurat improves upon previous versions by directly modeling the inherent mean-variance relationship in single-cell data. This method is implemented in the FindVariableFeatures() function, which, by default, selects 2,000 variable features per dataset. These features will then be used in downstream analyses, such as PCA.</p>
<pre><code class="language-R">pbmc &lt;- FindVariableFeatures(pbmc,
                selection.method = &quot;vst&quot;,
                nfeatures = 2000)

</code></pre>
<pre><code>Finding variable features for layer counts
</code></pre>
<pre><code class="language-R"># Identify the 10 most highly variable genes
top10 &lt;- head(VariableFeatures(pbmc), 10)
options(repr.plot.width=10, repr.plot.height= 6)

# plot variable features with and without labels
plot1 &lt;- VariableFeaturePlot(pbmc)
plot2 &lt;- LabelPoints(plot = plot1, points = top10, repel = TRUE)
print(plot1)
</code></pre>
<pre><code>When using repel, set xnudge and ynudge to 0 for optimal results
</code></pre>
<h3 id="scaling-the-data">Scaling the data</h3>
<p>Next, we apply a linear transformation (<code>scaling</code>) that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The <code>ScaleData()</code> function:</p>
<p>Shifts the expression of each gene, so that the mean expression across cells is 0
Scales the expression of each gene, so that the variance across cells is 1</p>
<p>This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate</p>
<p>The results of this are stored in <code>pbmc[["RNA"]]$scale.data</code></p>
<p>By default, only variable features are scaled.
You can specify the features argument to scale additional features.</p>
<pre><code class="language-R">all.genes &lt;- rownames(pbmc)
pbmc &lt;- ScaleData(pbmc, features = all.genes)
</code></pre>
<pre><code>Centering and scaling data matrix
</code></pre>
<h2 id="perform-linear-dimensional-reduction">Perform linear dimensional reduction</h2>
<pre><code class="language-R">pbmc &lt;- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
</code></pre>
<pre><code>PC_ 1 
Positive:  CD247, IL32, IL7R, RORA, CAMK4, LTB, INPP4B, STAT4, BCL2, ANK3 
       ZEB1, LEF1, TRBC1, CARD11, THEMIS, BACH2, MLLT3, RNF125, RASGRF2, NR3C2 
       NELL2, PDE3B, LINC01934, ENSG00000290067, PRKCA, TAFA1, PYHIN1, CTSW, CSGALNACT1, SAMD3 
Negative:  LYZ, FCN1, IRAK3, SLC8A1, CLEC7A, PLXDC2, IFI30, S100A9, SPI1, CYBB 
       MNDA, LRMDA, FGL2, VCAN, CTSS, RBM47, CSF3R, MCTP1, NCF2, TYMP 
       CYRIA, CST3, HCK, SLC11A1, WDFY3, S100A8, MS4A6A, MPEG1, LST1, CSTA 
PC_ 2 
Positive:  CD247, S100A4, STAT4, NKG7, CST7, CTSW, GZMA, SYTL3, RNF125, SAMD3 
       NCALD, MYO1F, MYBL1, KLRD1, PLCB1, TGFBR3, PRF1, GNLY, RAP1GAP2, RORA 
       CCL5, HOPX, FGFBP2, YES1, PYHIN1, FNDC3B, GNG2, SYNE1, KLRF1, SPON2 
Negative:  BANK1, MS4A1, CD79A, FCRL1, PAX5, IGHM, AFF3, LINC00926, NIBAN3, EBF1 
       IGHD, BLK, CD22, OSBPL10, HLA-DQA1, COL19A1, GNG7, KHDRBS2, RUBCNL, TNFRSF13C 
       COBLL1, RALGPS2, TCL1A, BCL11A, CDK14, CD79B, PLEKHG1, HLA-DQB1, IGKC, BLNK 
PC_ 3 
Positive:  TUBB1, GP9, GP1BB, PF4, CAVIN2, GNG11, NRGN, PPBP, RGS18, PRKAR2B 
       H2AC6, ACRBP, PTCRA, TMEM40, TREML1, CLU, LEF1, GPX1, CMTM5, SMANTIS 
       MPIG6B, CAMK4, MPP1, SPARC, ENSG00000289621, ITGB3, MYL9, MYL4, ITGA2B, F13A1 
Negative:  NKG7, CST7, GNLY, PRF1, KLRD1, GZMA, KLRF1, MCTP2, GZMB, FGFBP2 
       HOPX, SPON2, C1orf21, TGFBR3, VAV3, MYBL1, CTSW, SYNE1, NCALD, IL2RB 
       SAMD3, GNG2, BNC2, CEP78, YES1, RAP1GAP2, PDGFD, LINC02384, CARD11, CLIC3 
PC_ 4 
Positive:  CAMK4, INPP4B, IL7R, LEF1, PRKCA, PDE3B, MAML2, LTB, ANK3, PLCL1 
       BCL2, CDC14A, THEMIS, FHIT, NELL2, VIM, ENSG00000290067, MLLT3, TSHZ2, NR3C2 
       IL32, CMTM8, ENSG00000249806, ZEB1, SESN3, CSGALNACT1, TAFA1, LEF1-AS1, SLC16A10, LDLRAD4 
Negative:  GP1BB, GP9, TUBB1, PF4, CAVIN2, GNG11, PPBP, H2AC6, PTCRA, NRGN 
       ACRBP, TMEM40, PRKAR2B, RGS18, TREML1, MPIG6B, SMANTIS, CMTM5, CLU, SPARC 
       ITGA2B, ITGB3, ENSG00000289621, MYL9, CAPN1-AS1, MYL4, ENSG00000288758, DAB2, PDGFA-DT, CTTN 
PC_ 5 
Positive:  CDKN1C, HES4, FCGR3A, PELATON, CSF1R, IFITM3, SIGLEC10, TCF7L2, ZNF703, MS4A7 
       UICLM, ENSG00000287682, NEURL1, RHOC, FMNL2, CKB, FTL, CALHM6, HMOX1, BATF3 
       ACTB, MYOF, CCDC26, IFITM2, PAPSS2, RRAS, LST1, VMO1, SERPINA1, LRRC25 
Negative:  LINC02458, AKAP12, CA8, ENSG00000250696, SLC24A3, HDC, IL3RA, EPAS1, ENPP3, OSBPL1A 
       TRPM6, CCR3, CSF2RB, SEMA3C, THSD7A, ATP10D, DACH1, CRPPA, ATP8B4, TMEM164 
       ABHD5, CLC, CR1, ITGB8, LIN7A, TAFA2, MBOAT2, GATA2, DAPK2, GCSAML
</code></pre>
<p>You have several useful ways to visualize both cells and features that define the PCA, including <code>VizDimReduction()</code>, <code>DimPlot()</code>, and <code>DimHeatmap()</code>.</p>
<pre><code class="language-R">DimPlot(pbmc, reduction = &quot;pca&quot;) + NoLegend()
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_2024_files/BIOI611_scRNA_2024_60_0.png" /></p>
<p><code>DimHeatmap()</code> draws a heatmap focusing on a principal component. Both cells and genes are sorted by their principal component scores</p>
<pre><code class="language-R">DimHeatmap(pbmc, dims = 1:3, cells = 500, balanced = TRUE)
DimHeatmap(pbmc, dims = 20:22, cells = 500, balanced = TRUE)

</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_2024_files/BIOI611_scRNA_2024_62_0.png" /></p>
<p><img alt="png" src="../BIOI611_scRNA_2024_files/BIOI611_scRNA_2024_62_1.png" /></p>
<pre><code class="language-R">
</code></pre>
<h2 id="determine-the-dimensionality-of-the-dataset">Determine the ‘dimensionality’ of the dataset</h2>
<p>The elbow plot is a useful tool for determining the number of principal components (PCs) needed to capture the majority of variation in the data. It displays the standard deviation of each PC, with the "elbow" point typically serving as the threshold for selecting the most informative PCs. However, identifying the exact location of the elbow can be somewhat subjective.</p>
<pre><code class="language-R">ElbowPlot(pbmc,  ndims = 50)
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_2024_files/BIOI611_scRNA_2024_65_0.png" /></p>
<pre><code class="language-R"># Determine the percentage of variation associated with each PC
pct_var &lt;- pbmc[[&quot;pca&quot;]]@stdev / sum(pbmc[[&quot;pca&quot;]]@stdev) * 100

# Calculate cumulative percentages for each PC
cumu_pct &lt;- cumsum(pct_var)

# Identify the first PC where cumulative percentage exceeds 90% and individual variance is less than 5%
pc_number &lt;- which(cumu_pct &gt; 90 &amp; pct_var &lt; 5)[1]

pc_number
</code></pre>
<p>41</p>
<pre><code class="language-R">
</code></pre>
<h2 id="cluster-the-cells">Cluster the cells</h2>
<p>Seurat embeds cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected <code>quasi-cliques</code> or <code>communities</code>.</p>
<p>Seurat first constructs a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the <code>FindNeighbors()</code> function, and takes as input the previously defined dimensionality of the dataset.</p>
<p>To cluster the cells, Seurat next applies modularity optimization techniques such as the Louvain algorithm (default) or SLM [SLM, Blondel et al., Journal of Statistical Mechanics], to iteratively group cells together, with the goal of optimizing the standard modularity function. The FindClusters() function implements this procedure, and contains a resolution parameter that sets the <code>granularity</code> of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 1 typically returns good results for single-cell datasets of around 5k cells. Optimal resolution often increases for larger datasets. The clusters can be found using the <code>Idents()</code> function.</p>
<pre><code class="language-R">pbmc &lt;- FindNeighbors(pbmc, dims = 1:pc_number)
pbmc &lt;- FindClusters(pbmc, resolution = 0.1)
</code></pre>
<pre><code>Computing nearest neighbor graph

Computing SNN



Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 4559
Number of edges: 184776

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9713
Number of communities: 10
Elapsed time: 0 seconds
</code></pre>
<pre><code class="language-R"># Look at cluster IDs of the first 5 cells
head(Idents(pbmc), 5)
</code></pre>
<style>
.dl-inline {width: auto; margin:0; padding: 0}
.dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}
.dl-inline>dt::after {content: ":\0020"; padding-right: .5ex}
.dl-inline>dt:not(:first-of-type) {padding-left: .5ex}
</style>
<dl class=dl-inline><dt>AAACCCATCAGATGCT-1</dt><dd>1</dd><dt>AAACGAAAGTGCTACT-1</dt><dd>0</dd><dt>AAACGAAGTCGTAATC-1</dt><dd>6</dd><dt>AAACGAAGTTGCCAAT-1</dt><dd>5</dd><dt>AAACGAATCCGAGGCT-1</dt><dd>4</dd></dl>

<details>
    <summary style=display:list-item;cursor:pointer>
        <strong>Levels</strong>:
    </summary>
    <style>
    .list-inline {list-style: none; margin:0; padding: 0}
    .list-inline>li {display: inline-block}
    .list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
    </style>
    <ol class=list-inline><li>'0'</li><li>'1'</li><li>'2'</li><li>'3'</li><li>'4'</li><li>'5'</li><li>'6'</li><li>'7'</li><li>'8'</li><li>'9'</li></ol>
</details>

<h2 id="run-non-linear-dimensional-reduction-umaptsne">Run non-linear dimensional reduction (UMAP/tSNE)</h2>
<p>To visualize and explore these datasets, Seurat  offers several non-linear dimensional reduction techniques, such as tSNE and UMAP.</p>
<p>The goal of tSNE/UMAP is to learn underlying structure in the dataset, in order to place similar cells together in low-dimensional space. Therefore, cells that are grouped together within graph-based clusters determined above should co-localize on these dimension reduction plots.</p>
<pre><code class="language-R">pbmc &lt;- RunUMAP(pbmc, dims = 1:pc_number)

</code></pre>
<pre><code>Warning message:
“The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
This message will be shown once per session”
02:09:13 UMAP embedding parameters a = 0.9922 b = 1.112

02:09:13 Read 4559 rows and found 41 numeric columns

02:09:13 Using Annoy for neighbor search, n_neighbors = 30

02:09:13 Building Annoy index with metric = cosine, n_trees = 50

0%   10   20   30   40   50   60   70   80   90   100%

[----|----|----|----|----|----|----|----|----|----|

*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
|

02:09:14 Writing NN index file to temp file /tmp/RtmpRPPsRm/file52075ef52dee

02:09:14 Searching Annoy index using 1 thread, search_k = 3000

02:09:16 Annoy recall = 100%

02:09:17 Commencing smooth kNN distance calibration using 1 thread
 with target n_neighbors = 30

02:09:20 Found 2 connected components, 
falling back to 'spca' initialization with init_sdev = 1

02:09:20 Using 'irlba' for PCA

02:09:20 PCA: 2 components explained 46.09% variance

02:09:20 Scaling init to sdev = 1

02:09:20 Commencing optimization for 500 epochs, with 188320 positive edges

02:09:28 Optimization finished
</code></pre>
<pre><code class="language-R">DimPlot(pbmc, reduction = &quot;umap&quot;)
</code></pre>
<p><img alt="png" src="../BIOI611_scRNA_2024_files/BIOI611_scRNA_2024_73_0.png" /></p>
<h2 id="finding-differentially-expressed-features-cluster-biomarkers">Finding differentially expressed features (cluster biomarkers)</h2>
<pre><code class="language-R"># find markers for every cluster compared to all remaining cells, report only the positive
# ones
pbmc.markers &lt;- FindAllMarkers(pbmc, only.pos = TRUE)
pbmc.markers %&gt;%
    group_by(cluster) %&gt;%
    dplyr::filter(avg_log2FC &gt; 1)
</code></pre>
<pre><code>Calculating cluster 0

Calculating cluster 1

Calculating cluster 2

Calculating cluster 3

Calculating cluster 4

Calculating cluster 5

Calculating cluster 6

Calculating cluster 7

Calculating cluster 8

Calculating cluster 9
</code></pre>
<table class="dataframe">
<caption>A grouped_df: 14138 × 7</caption>
<thead>
    <tr><th scope=col>p_val</th><th scope=col>avg_log2FC</th><th scope=col>pct.1</th><th scope=col>pct.2</th><th scope=col>p_val_adj</th><th scope=col>cluster</th><th scope=col>gene</th></tr>
    <tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
    <tr><td> 0.000000e+00</td><td>2.193938</td><td>0.974</td><td>0.426</td><td> 0.000000e+00</td><td>0</td><td>IL32    </td></tr>
    <tr><td>3.852209e-275</td><td>1.679326</td><td>0.951</td><td>0.362</td><td>9.547700e-271</td><td>0</td><td>CD3D    </td></tr>
    <tr><td>9.596093e-260</td><td>1.803885</td><td>0.888</td><td>0.352</td><td>2.378392e-255</td><td>0</td><td>TRAC    </td></tr>
    <tr><td>4.781243e-252</td><td>1.573425</td><td>0.959</td><td>0.532</td><td>1.185031e-247</td><td>0</td><td>SYNE2   </td></tr>
    <tr><td>3.690404e-235</td><td>1.512388</td><td>0.917</td><td>0.348</td><td>9.146665e-231</td><td>0</td><td>CD3G    </td></tr>
    <tr><td>9.967907e-232</td><td>1.809323</td><td>0.910</td><td>0.485</td><td>2.470546e-227</td><td>0</td><td>RORA    </td></tr>
    <tr><td>1.117947e-227</td><td>1.378699</td><td>0.951</td><td>0.380</td><td>2.770831e-223</td><td>0</td><td>CD3E    </td></tr>
    <tr><td>1.039791e-208</td><td>2.253821</td><td>0.681</td><td>0.233</td><td>2.577123e-204</td><td>0</td><td>NIBAN1  </td></tr>
    <tr><td>2.027326e-187</td><td>1.595996</td><td>0.810</td><td>0.357</td><td>5.024728e-183</td><td>0</td><td>CD2     </td></tr>
    <tr><td>8.928063e-181</td><td>1.512822</td><td>0.846</td><td>0.383</td><td>2.212820e-176</td><td>0</td><td>IL7R    </td></tr>
    <tr><td>2.768362e-179</td><td>2.583397</td><td>0.463</td><td>0.091</td><td>6.861385e-175</td><td>0</td><td>PRDM1   </td></tr>
    <tr><td>2.609892e-166</td><td>1.825963</td><td>0.777</td><td>0.389</td><td>6.468617e-162</td><td>0</td><td>PPP1R16B</td></tr>
    <tr><td>8.007361e-166</td><td>3.888615</td><td>0.310</td><td>0.031</td><td>1.984624e-161</td><td>0</td><td>TRGC2   </td></tr>
    <tr><td>1.861258e-163</td><td>1.694587</td><td>0.717</td><td>0.296</td><td>4.613128e-159</td><td>0</td><td>CD6     </td></tr>
    <tr><td>1.403718e-158</td><td>1.072541</td><td>0.969</td><td>0.807</td><td>3.479115e-154</td><td>0</td><td>PPP2R5C </td></tr>
    <tr><td>9.430895e-158</td><td>2.200303</td><td>0.525</td><td>0.147</td><td>2.337447e-153</td><td>0</td><td>CD5     </td></tr>
    <tr><td>5.325382e-153</td><td>1.314341</td><td>0.835</td><td>0.433</td><td>1.319896e-148</td><td>0</td><td>SPOCK2  </td></tr>
    <tr><td>5.541697e-151</td><td>1.501378</td><td>0.757</td><td>0.363</td><td>1.373510e-146</td><td>0</td><td>GPRIN3  </td></tr>
    <tr><td>1.783366e-148</td><td>1.524168</td><td>0.815</td><td>0.492</td><td>4.420072e-144</td><td>0</td><td>LRRC8C  </td></tr>
    <tr><td>5.511823e-146</td><td>1.130197</td><td>0.937</td><td>0.772</td><td>1.366105e-141</td><td>0</td><td>EML4    </td></tr>
    <tr><td>4.406921e-145</td><td>1.807049</td><td>0.821</td><td>0.535</td><td>1.092255e-140</td><td>0</td><td>TNFAIP3 </td></tr>
    <tr><td>8.596339e-145</td><td>1.322956</td><td>0.800</td><td>0.391</td><td>2.130603e-140</td><td>0</td><td>LIME1   </td></tr>
    <tr><td>1.601435e-143</td><td>1.301753</td><td>0.861</td><td>0.505</td><td>3.969156e-139</td><td>0</td><td>ATXN1   </td></tr>
    <tr><td>3.308730e-140</td><td>4.011786</td><td>0.283</td><td>0.033</td><td>8.200689e-136</td><td>0</td><td>GZMK    </td></tr>
    <tr><td>1.881698e-138</td><td>1.586776</td><td>0.770</td><td>0.413</td><td>4.663787e-134</td><td>0</td><td>PHACTR2 </td></tr>
    <tr><td>8.271921e-138</td><td>1.486293</td><td>0.791</td><td>0.479</td><td>2.050196e-133</td><td>0</td><td>TTC39C  </td></tr>
    <tr><td>8.346144e-135</td><td>3.699470</td><td>0.245</td><td>0.020</td><td>2.068592e-130</td><td>0</td><td>MIAT    </td></tr>
    <tr><td>2.291535e-126</td><td>1.420132</td><td>0.725</td><td>0.348</td><td>5.679570e-122</td><td>0</td><td>ANK3    </td></tr>
    <tr><td>2.583870e-126</td><td>2.075113</td><td>0.477</td><td>0.151</td><td>6.404121e-122</td><td>0</td><td>PBX4    </td></tr>
    <tr><td>3.042390e-125</td><td>1.333502</td><td>0.694</td><td>0.316</td><td>7.540564e-121</td><td>0</td><td>OPTN    </td></tr>
    <tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>
    <tr><td>0.009116629</td><td>3.631680</td><td>0.021</td><td>0.002</td><td>1</td><td>9</td><td>H4C1           </td></tr>
    <tr><td>0.009116629</td><td>3.608558</td><td>0.021</td><td>0.002</td><td>1</td><td>9</td><td>ENSG00000289291</td></tr>
    <tr><td>0.009116629</td><td>3.605235</td><td>0.021</td><td>0.002</td><td>1</td><td>9</td><td>VAMP1-AS1      </td></tr>
    <tr><td>0.009116629</td><td>3.591495</td><td>0.021</td><td>0.002</td><td>1</td><td>9</td><td>ENSG00000249328</td></tr>
    <tr><td>0.009116629</td><td>3.567020</td><td>0.021</td><td>0.002</td><td>1</td><td>9</td><td>ERICH2-DT      </td></tr>
    <tr><td>0.009116629</td><td>3.550000</td><td>0.021</td><td>0.002</td><td>1</td><td>9</td><td>NLRP10         </td></tr>
    <tr><td>0.009116629</td><td>3.518027</td><td>0.021</td><td>0.002</td><td>1</td><td>9</td><td>ENSG00000270087</td></tr>
    <tr><td>0.009116629</td><td>3.496166</td><td>0.021</td><td>0.002</td><td>1</td><td>9</td><td>ENSG00000253593</td></tr>
    <tr><td>0.009116629</td><td>3.393207</td><td>0.021</td><td>0.002</td><td>1</td><td>9</td><td>ADAM11         </td></tr>
    <tr><td>0.009116629</td><td>3.272524</td><td>0.021</td><td>0.002</td><td>1</td><td>9</td><td>ENSG00000228150</td></tr>
    <tr><td>0.009116629</td><td>3.266281</td><td>0.021</td><td>0.002</td><td>1</td><td>9</td><td>LINC03065      </td></tr>
    <tr><td>0.009116629</td><td>3.221456</td><td>0.021</td><td>0.002</td><td>1</td><td>9</td><td>STARD13-AS     </td></tr>
    <tr><td>0.009116629</td><td>3.109640</td><td>0.021</td><td>0.002</td><td>1</td><td>9</td><td>FGGY-DT        </td></tr>
    <tr><td>0.009116629</td><td>2.116185</td><td>0.021</td><td>0.002</td><td>1</td><td>9</td><td>ENSG00000285751</td></tr>
    <tr><td>0.009353183</td><td>1.486986</td><td>0.146</td><td>0.057</td><td>1</td><td>9</td><td>TRPM2          </td></tr>
    <tr><td>0.009560927</td><td>1.198107</td><td>0.083</td><td>0.024</td><td>1</td><td>9</td><td>SYCP3          </td></tr>
    <tr><td>0.009563781</td><td>1.367524</td><td>0.062</td><td>0.015</td><td>1</td><td>9</td><td>MAP7           </td></tr>
    <tr><td>0.009595402</td><td>1.283987</td><td>0.083</td><td>0.024</td><td>1</td><td>9</td><td>PPP1R13L       </td></tr>
    <tr><td>0.009686754</td><td>2.584749</td><td>0.042</td><td>0.008</td><td>1</td><td>9</td><td>PRRG2          </td></tr>
    <tr><td>0.009706714</td><td>2.407445</td><td>0.042</td><td>0.008</td><td>1</td><td>9</td><td>LINC02185      </td></tr>
    <tr><td>0.009746744</td><td>2.473951</td><td>0.042</td><td>0.008</td><td>1</td><td>9</td><td>LINC02901      </td></tr>
    <tr><td>0.009766814</td><td>2.300928</td><td>0.042</td><td>0.008</td><td>1</td><td>9</td><td>WNK3           </td></tr>
    <tr><td>0.009766814</td><td>1.640083</td><td>0.042</td><td>0.008</td><td>1</td><td>9</td><td>CALCRL-AS1     </td></tr>
    <tr><td>0.009786921</td><td>2.434878</td><td>0.042</td><td>0.008</td><td>1</td><td>9</td><td>ENSG00000272112</td></tr>
    <tr><td>0.009786922</td><td>2.438294</td><td>0.042</td><td>0.008</td><td>1</td><td>9</td><td>ENSG00000277589</td></tr>
    <tr><td>0.009847464</td><td>2.262934</td><td>0.042</td><td>0.008</td><td>1</td><td>9</td><td>PCDH15         </td></tr>
    <tr><td>0.009847464</td><td>2.064514</td><td>0.042</td><td>0.008</td><td>1</td><td>9</td><td>CIBAR1         </td></tr>
    <tr><td>0.009888011</td><td>2.173570</td><td>0.042</td><td>0.008</td><td>1</td><td>9</td><td>SEZ6           </td></tr>
    <tr><td>0.009908340</td><td>1.746860</td><td>0.042</td><td>0.008</td><td>1</td><td>9</td><td>RTKN           </td></tr>
    <tr><td>0.009993542</td><td>1.340809</td><td>0.146</td><td>0.059</td><td>1</td><td>9</td><td>ENSG00000287100</td></tr>
</tbody>
</table>

<pre><code class="language-R"># find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers &lt;- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3))
head(cluster5.markers, n = 5)
</code></pre>
<table class="dataframe">
<caption>A data.frame: 5 × 5</caption>
<thead>
    <tr><th></th><th scope=col>p_val</th><th scope=col>avg_log2FC</th><th scope=col>pct.1</th><th scope=col>pct.2</th><th scope=col>p_val_adj</th></tr>
    <tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
    <tr><th scope=row>PTPRK</th><td>4.438261e-196</td><td>5.477436</td><td>0.594</td><td>0.020</td><td>1.100023e-191</td></tr>
    <tr><th scope=row>NRCAM</th><td>9.537544e-196</td><td>7.815745</td><td>0.496</td><td>0.005</td><td>2.363880e-191</td></tr>
    <tr><th scope=row>AIF1</th><td>1.738908e-192</td><td>4.287096</td><td>0.804</td><td>0.077</td><td>4.309884e-188</td></tr>
    <tr><th scope=row>LINC02446</th><td>3.824707e-176</td><td>4.782107</td><td>0.638</td><td>0.039</td><td>9.479537e-172</td></tr>
    <tr><th scope=row>NELL2</th><td>7.816637e-150</td><td>3.058779</td><td>0.969</td><td>0.213</td><td>1.937354e-145</td></tr>
</tbody>
</table>

<pre><code class="language-R">VlnPlot(pbmc, features = c(&quot;PTPRK&quot;, &quot;NRCAM&quot;))

</code></pre>
<pre><code>Error: Can't find method for generic `&amp;(e1, e2)`:
- e1: &lt;patchwork&gt;
- e2: &lt;theme&gt;
Traceback:


1. ExIPlot(object = object, type = ifelse(test = split.plot, yes = "splitViolin", 
 .     no = "violin"), features = features, idents = idents, ncol = ncol, 
 .     sort = sort, assay = assay, y.max = y.max, same.y.lims = same.y.lims, 
 .     adjust = adjust, pt.size = pt.size, alpha = alpha, cols = cols, 
 .     group.by = group.by, split.by = split.by, log = log, layer = layer, 
 .     stack = stack, combine = combine, fill.by = fill.by, flip = flip, 
 .     add.noise = add.noise, raster = raster)

2. Ops.S7_object(plots, NoLegend())

3. stop(cnd)
</code></pre>
<pre><code class="language-R">VlnPlot(pbmc, features = c(&quot;NKG7&quot;, &quot;PF4&quot;), slot = &quot;counts&quot;, log = TRUE)

</code></pre>
<pre><code class="language-R">FeaturePlot(pbmc, features = c(&quot;MS4A1&quot;, &quot;GNLY&quot;, &quot;CD3E&quot;, &quot;CD14&quot;, &quot;FCER1A&quot;, &quot;FCGR3A&quot;, &quot;LYZ&quot;, &quot;PPBP&quot;,
    &quot;CD8A&quot;))
</code></pre>
<pre><code class="language-R">pbmc.markers %&gt;%
    group_by(cluster) %&gt;%
    dplyr::filter(avg_log2FC &gt; 1) %&gt;%
    slice_head(n = 10) %&gt;%
    ungroup() -&gt; top10
DoHeatmap(pbmc, features = top10$gene) + NoLegend()
</code></pre>
<h2 id="cell-type-annotation-using-singler">Cell type annotation using <code>SingleR</code></h2>
<pre><code class="language-R">library(SingleCellExperiment )
</code></pre>
<pre><code class="language-R">sce &lt;- as.SingleCellExperiment(pbmc)
sce &lt;- scater::logNormCounts(sce)

</code></pre>
<pre><code class="language-R">sce
</code></pre>
<pre><code class="language-R"># Download and cache the normalized expression values of the data
# stored in the Human Primary Cell Atlas. The data will be
#  downloaded from ExperimentHub, returning a SummarizedExperiment
# object for further use.
hpca &lt;- HumanPrimaryCellAtlasData()

# Obtain human bulk RNA-seq data from Blueprint and ENCODE
blueprint &lt;- BlueprintEncodeData()

</code></pre>
<pre><code class="language-R">pred.hpca &lt;- SingleR(test = sce,
      ref = hpca, labels = hpca$label.main)
tab_hpca &lt;- table(pred.hpca$pruned.labels)
write.csv(sort(tab_hpca, decreasing=TRUE), 'pbmc_annotations_HPCA_general.csv', row.names=FALSE)
</code></pre>
<pre><code class="language-R">tab_hpca
</code></pre>
<p>Each row of the output DataFrame contains prediction results for a single cell. Labels are shown before (labels) and after pruning (pruned.labels), along with the associated scores.</p>
<pre><code class="language-R">head(pred.hpca)
</code></pre>
<pre><code class="language-R">pred.blueprint &lt;- SingleR(test = sce,
      ref = blueprint, labels = blueprint$label.main)
tab_blueprint &lt;- table(pred.blueprint$pruned.labels)
  write.csv(sort(tab_blueprint, decreasing=TRUE), 'pbmc_annotations_BlueprintENCODE_general.csv', row.names=FALSE)
</code></pre>
<pre><code class="language-R">tab_blueprint
</code></pre>
<pre><code class="language-R">head(pred.blueprint)
</code></pre>
<pre><code class="language-R">pbmc$singleR_hpca = pred.hpca$pruned.labels
pbmc$singleR_blueprint = pred.blueprint$pruned.labels
</code></pre>
<pre><code class="language-R">Idents(pbmc) = pbmc$singleR_hpca
DimPlot(pbmc, reduction = &quot;umap&quot;,
         label = TRUE, pt.size = 0.5,
          repel = TRUE) + NoLegend()
# Change back to cluster snn_res.0.1
Idents(pbmc) = pbmc$`RNA_snn_res.0.1`
</code></pre>
<pre><code class="language-R">Idents(pbmc) = pbmc$singleR_blueprint
DimPlot(pbmc, reduction = &quot;umap&quot;,
         label = TRUE, pt.size = 0.5,
          repel = TRUE) + NoLegend()
# Change back to cluster snn_res.0.1
Idents(pbmc) = pbmc$`RNA_snn_res.0.1`
</code></pre>
<pre><code class="language-R">
</code></pre>
<h2 id="manual-annotation">Manual annotation</h2>
<table>
<thead>
<tr>
<th>Markers</th>
<th>Cell Type</th>
</tr>
</thead>
<tbody>
<tr>
<td>IL7R, CCR7</td>
<td>Naive CD4+ T</td>
</tr>
<tr>
<td>CD14, LYZ</td>
<td>CD14+ Mono</td>
</tr>
<tr>
<td>IL7R, S100A4</td>
<td>Memory CD4+</td>
</tr>
<tr>
<td>MS4A1</td>
<td>B</td>
</tr>
<tr>
<td>CD8A</td>
<td>CD8+ T</td>
</tr>
<tr>
<td>FCGR3A, MS4A7</td>
<td>FCGR3A+ Mono</td>
</tr>
<tr>
<td>GNLY, NKG7</td>
<td>NK</td>
</tr>
<tr>
<td>FCER1A, CST3</td>
<td>DC</td>
</tr>
<tr>
<td>PPBP</td>
<td>Platelet</td>
</tr>
</tbody>
</table>
<pre><code class="language-R">table(Idents(pbmc))
</code></pre>
<pre><code class="language-R">
</code></pre>
<pre><code class="language-R">FeaturePlot(pbmc, features = c(&quot;IL7R&quot;, &quot;CD14&quot;, &quot;CCR7&quot;, &quot;S100A4&quot;, &quot;MS4A1&quot;))
</code></pre>
<pre><code class="language-R">DimPlot(pbmc, reduction = &quot;umap&quot;,
         label = TRUE, pt.size = 0.5,
          repel = TRUE) + NoLegend()
</code></pre>
<pre><code class="language-R">pbmc = RenameIdents(pbmc, &quot;2&quot;=&quot;CD14+ Mono&quot;)
pbmc = RenameIdents(pbmc, &quot;1&quot;=&quot;Naive CD4+ T&quot;)
pbmc = RenameIdents(pbmc, &quot;0&quot;=&quot;Memory CD4+&quot;)
### Add more manual annotation based on your checking

</code></pre>
<pre><code class="language-R">DimPlot(pbmc, reduction = &quot;umap&quot;,
         label = TRUE, pt.size = 0.5,
          repel = TRUE) + NoLegend()
</code></pre>
<h2 id="save-the-seurat-object">Save the Seurat object</h2>
<pre><code class="language-R">saveRDS(pbmc, file = &quot;Seurat_object_pbmc_final.rds&quot;)
</code></pre>
<pre><code class="language-R">remotes::install_github(&quot;10xGenomics/loupeR&quot;)
loupeR::setup()
</code></pre>
<pre><code class="language-R">library(loupeR)
</code></pre>
<pre><code class="language-R">create_loupe_from_seurat(pbmc, output_name = &quot;Seurat_object_pbmc_cloupe&quot;)

</code></pre>
<pre><code class="language-R">sessionInfo()
</code></pre>
<h2 id="reference">Reference</h2>
<p>https://monashbioinformaticsplatform.github.io/Single-Cell-Workshop/pbmc3k_tutorial.html</p>
<p>https://bioinformatics.ccr.cancer.gov/docs/getting-started-with-scrna-seq/IntroToR_Seurat/</p>
<p>https://hbctraining.github.io/scRNA-seq/lessons/elbow_plot_metric.html</p>
              
            </div>
          </div><footer>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>Copyright &copy; 2025 Shaojun Xie</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/Bix4UMD/BIOI611_lab.git" class="fa fa-code-fork" style="color: #fcfcfc"> Bix4UMD/BIOI611_lab</a>
        </span>
    
    
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../javascripts/mathjax.js"></script>
      <script src="https://unpkg.com/mathjax@3/es5/tex-mml-chtml.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
