<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Shaojun Xie" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Initial analysis of 10x scRNA-seq data for C. elegans using cellranger - Lab note for UMD BIOI611</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Initial analysis of 10x scRNA-seq data for C. elegans using cellranger";
        var mkdocs_page_input_path = "BIOI611_scRNA_seq_cele_cellranger.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/yaml.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/rust.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> Lab note for UMD BIOI611
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption"><span class="caption-text">Get ready for the course</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../basic_linux/">Basic Linux</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Bulk RNA-seq</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../bulkRNAseq_lab/">Read QC and alignment</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Phred_FQ/">Phred score in FASTQ</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_introR/">Minimal R Introduction</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_DESeq2_analysis/">DEG analysis using DESeq2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_DAVID/">Pathway analysis using DAVID</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_ballgown_example/">DE analysis at isoform-leve using ballgown (example data)</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_bulkRNA_SE_ballgown/">DE analysis at isoform-leve using ballgown (C. ele data)</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_long_read_transcriptome/">Long read transcriptome</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">scRNA-seq</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_analysis_10x_dataset_PBMC/">Initial analysis of 10x scRNA-seq data for human PBMC using cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_scRNA/">Downstream analysis of 10x scRNA-seq data for human PBMC using Seurat</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">Initial analysis of 10x scRNA-seq data for C. elegans using cellranger</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#data-availability">Data availability</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#prepare-the-genome">Prepare the genome</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#run-cellranger-count">Run cellranger count</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#aggregate-the-cellranger-count-results">Aggregate the cellranger count results</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BIOI611_scRNA_cele/">Downstream analysis of 10x scRNA-seq data for C. elegans data</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bioi611_monocle_cele/">Downstream analysis - trajectory analysis using Monocle3</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">Lab note for UMD BIOI611</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">scRNA-seq</li>
      <li class="breadcrumb-item active">Initial analysis of 10x scRNA-seq data for C. elegans using cellranger</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/Bix4UMD/BIOI611_lab.git/edit/master/docs/BIOI611_scRNA_seq_cele_cellranger.md">Edit on Bix4UMD/BIOI611_lab</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <p><a href="https://colab.research.google.com/github/Bix4UMD/BIOI611_lab/blob/main/docs/BIOI611_scRNA_seq_cele_cellranger.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<h1 id="_1"></h1>
<h2 id="data-availability">Data availability</h2>
<p>Data can be obtained from the link below:
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE126954</p>
<ul>
<li>GSM3618670    UW synchronized 300 min post bleach</li>
<li>GSM3618671    UW synchronized 400 min post bleach</li>
<li>GSM3618672    UW synchronized 500 min post bleach batch 1</li>
<li>GSM3618673    UW synchronized 500 min post bleach batch 2</li>
<li>GSM3618674    UPenn mixed embryo batch r17</li>
<li>GSM3618675    UPenn mixed embryo batch b01</li>
<li>GSM3618676    UPenn mixed embryo batch b02</li>
</ul>
<p>In this class, you will only work on: <code>300 min post bleach</code>, <code>400 min post bleach</code> and <code>500 min post bleach batch 1</code></p>
<p>An copy of the data has been stored here:
/scratch/zt1/project/bioi611/shared/raw_data/10x_cele_data/scRNA/</p>
<ol>
<li>If you want to download and prepare the files yourself,</li>
</ol>
<p>Here is the process:</p>
<p>For each sample, fetch the sra files using <code>prefectch</code></p>
<p>For example:</p>
<pre><code>export PATH=/scratch/zt1/project/bioi611/shared/software/sratoolkit.3.1.1-centos_linux64/bin:$PATH
prefetch SRR8611967
</code></pre>
<ol>
<li>Convert sra file to fastq files</li>
</ol>
<pre><code>fasterq-dump  --outdir &lt;output_folder&gt;  --include-technical --split-files &lt;sra_file&gt;
</code></pre>
<h2 id="prepare-the-genome">Prepare the genome</h2>
<p>You don't need to run this step. The content in this part is to show you if you want to prepare the reference genome for cellranger, how you can prepare.</p>
<pre><code>## /scratch/zt1/project/bioi611/shared/reference/cellranger_mkref/
$ cat  /scratch/zt1/project/bioi611/shared/reference/cellranger_mkref/scRNA_cellranger_mkref.sub
#!/bin/bash
#SBATCH --partition=standard
#SBATCH -t 40:00:00
#SBATCH -n 1
#SBATCH -c 26
#SBATCH --mem=250g
#SBATCH --job-name=scRNA_cellranger_mkref
#SBATCH --mail-type=FAIL,BEGIN,END
#SBATCH --error=%x-%J-%u.err
#SBATCH --output=%x-%J-%u.out
export PATH=/scratch/zt1/project/bioi611/shared/software/cellranger-8.0.1/bin:$PATH
cellranger mkgtf ../Caenorhabditis_elegans.WBcel235.111.gtf \
                 Caenorhabditis_elegans.WBcel235.111.filtered.gtf \
                 --attribute=gene_biotype:protein_coding &gt; scRNA_cellranger_mkref.filter_gtf.log 2&gt;&amp;1
cellranger mkref --genome=Caenorhabditis_elegans_genome \
                 --fasta=../Caenorhabditis_elegans.WBcel235.dna.toplevel.fa \
                 --genes=Caenorhabditis_elegans.WBcel235.111.filtered.gtf \
                 &gt; scRNA_cellranger_mkref.log 2&gt;&amp;1  
</code></pre>
<h2 id="run-cellranger-count">Run <code>cellranger count</code></h2>
<pre><code>sbatch /scratch/zt1/project/bioi611/shared/scripts/scRNA_10x_cele_cellranger_count.Uwsync_300min.sub
sbatch /scratch/zt1/project/bioi611/shared/scripts/scRNA_10x_cele_cellranger_count.Uwsync_400min.sub
sbatch /scratch/zt1/project/bioi611/shared/scripts/scRNA_10x_cele_cellranger_count.Uwsync_500min.sub
</code></pre>
<h2 id="aggregate-the-cellranger-count-results">Aggregate the <code>cellranger count</code> results</h2>
<p>Many experiments generate data for multiple samples. Depending on the experimental design, these samples may represent replicates from the same set of cells, cells from different tissues or time points from the same individual, or cells from different individuals. These samples could be processed through various Gel Bead-in Emulsion (GEM) wells wells or multiplexed within the same GEM well on Chromium instruments.</p>
<p>To work with data from multiple GEM wells, you can aggregate and analyze the outputs from multiple runs of each of these pipelines using <code>cellranger aggr</code>.</p>
<pre><code>sbatch /scratch/zt1/project/bioi611/shared/scripts/scRNA_10x_cele_cellranger_aggr.sub
</code></pre>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../BIOI611_scRNA/" class="btn btn-neutral float-left" title="Downstream analysis of 10x scRNA-seq data for human PBMC using Seurat"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../BIOI611_scRNA_cele/" class="btn btn-neutral float-right" title="Downstream analysis of 10x scRNA-seq data for C. elegans data">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>Copyright &copy; 2025 Shaojun Xie</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/Bix4UMD/BIOI611_lab.git" class="fa fa-code-fork" style="color: #fcfcfc"> Bix4UMD/BIOI611_lab</a>
        </span>
    
    
      <span><a href="../BIOI611_scRNA/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../BIOI611_scRNA_cele/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../javascripts/mathjax.js"></script>
      <script src="https://unpkg.com/mathjax@3/es5/tex-mml-chtml.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
