{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Bix4UMD/BIOI611_lab/blob/main/docs/BIOI611_DESeq2_analysis.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Analysis of RNA-seq data using R"
      ],
      "metadata": {
        "id": "_V8RbUWl4EQL"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Navigation in the file system and read the count files\n"
      ],
      "metadata": {
        "id": "2Ao8YyE7OCaP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "getwd()"
      ],
      "metadata": {
        "id": "8CDHA4SsOLVl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "list.files()"
      ],
      "metadata": {
        "id": "cXZ-B7hvOOy7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Define base URL and files to download\n",
        "base_url <- \"https://raw.githubusercontent.com/Bix4UMD/BIOI611_lab/refs/heads/main/data/bulkRNA_SE_tab/\"\n",
        "files <- c(\"N2_day1_rep1.ReadsPerGene.out.tab\",\n",
        "           \"N2_day1_rep2.ReadsPerGene.out.tab\",\n",
        "           \"N2_day1_rep3.ReadsPerGene.out.tab\",\n",
        "           \"N2_day7_rep1.ReadsPerGene.out.tab\",\n",
        "           \"N2_day7_rep2.ReadsPerGene.out.tab\",\n",
        "           \"N2_day7_rep3.ReadsPerGene.out.tab\"\n",
        "           )\n",
        "\n",
        "# Download each file\n",
        "for (f in files) {\n",
        "  download.file(paste0(base_url, f), destfile = f, method = \"libcurl\")\n",
        "}"
      ],
      "metadata": {
        "id": "mdXA17IIQ1Qc"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "list.files()"
      ],
      "metadata": {
        "id": "1f2lksAfRi7B"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "file_paths <- list.files(pattern = \"*.ReadsPerGene.out.tab\")\n",
        "file_paths"
      ],
      "metadata": {
        "id": "jIxOGN2TJA9l"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Function to read the STAR ReadsPerGene.out.tab file\n",
        "read_star_file <- function(file_path) {\n",
        "  # Read the file\n",
        "  df <- read.table(file_path, header = FALSE, stringsAsFactors = FALSE)\n",
        "\n",
        "  # Keep only the first (gene) and second (unstranded counts) columns\n",
        "  library(dplyr)\n",
        "  df <- df %>% select(V1, V2)\n",
        "\n",
        "  # Rename the columns for clarity (GeneID and counts for this sample)\n",
        "  colnames(df) <- c(\"GeneID\", gsub(\".ReadsPerGene.out.tab\", \"\", basename(file_path)))\n",
        "\n",
        "  return(df)\n",
        "}\n",
        "\n",
        "# Read all files into a list of data frames\n",
        "list_of_dfs <- lapply(file_paths, read_star_file)\n",
        "\n",
        "# Merge all data frames by the GeneID column\n",
        "merged_df <- Reduce(function(x, y) merge(x, y, by = \"GeneID\"), list_of_dfs)\n",
        "\n",
        "merged_df <- merged_df[-c(1:4), ]\n",
        "\n",
        "# Check the first few rows of the combined data frame\n",
        "head(merged_df)\n",
        "\n",
        "# Optionally, write the combined data frame to a CSV file\n",
        "write.csv(merged_df, \"combined_gene_counts.csv\", row.names = FALSE)"
      ],
      "metadata": {
        "id": "rm810m88Ir7C"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "class(list_of_dfs)"
      ],
      "metadata": {
        "id": "Ck_KDuTRKPjc"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "head(list_of_dfs[[2]])"
      ],
      "metadata": {
        "id": "3543cqgFKf-x"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "8_d9-qKKK0dq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "rownames(merged_df) = merged_df$GeneID\n"
      ],
      "metadata": {
        "id": "Tj_Fg6CZKi3a"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "head(merged_df)"
      ],
      "metadata": {
        "id": "wExTyfPTLig0"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# NULL reserved word representing empty\n",
        "merged_df$GeneID = NULL\n",
        "head(merged_df)"
      ],
      "metadata": {
        "id": "oKzfrU3hLsFh"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "subset_df4test <- merged_df[, c(\"N2_day1_rep1\",\t\"N2_day1_rep2\", \"N2_day1_rep3\")]\n",
        "head(subset_df4test)"
      ],
      "metadata": {
        "id": "xmKJQe3sNXAk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Check count matrix"
      ],
      "metadata": {
        "id": "UrPiC3BqOXyY"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Different samples have different total number of counts\n"
      ],
      "metadata": {
        "id": "kM9lGnPZObfi"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "as.data.frame(colSums(merged_df))"
      ],
      "metadata": {
        "id": "CHNSWb0aYRxg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "barplot(colSums(merged_df),\n",
        "         las = 2,\n",
        "         cex.names= 0.6)"
      ],
      "metadata": {
        "id": "a4hIQoxESLQf"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# inkscape: an open source software for editing the graph saved in pdf\n",
        "pdf(\"total_count_barplot.pdf\")\n",
        "barplot(colSums(merged_df),\n",
        "         las = 2,\n",
        "         cex.names= 0.6)\n",
        "dev.off()"
      ],
      "metadata": {
        "id": "JWehBTppO1Mt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "coldata <- colnames(merged_df)\n",
        "coldata_df <- cbind(group = gsub(\"_rep\\\\d\", \"\", coldata))\n",
        "coldata_df"
      ],
      "metadata": {
        "id": "ffuI0idTI1NJ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "rownames(coldata_df) = coldata\n",
        "coldata_df"
      ],
      "metadata": {
        "id": "r2ugw1F_TEza"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Instsall required R packages\n"
      ],
      "metadata": {
        "id": "leS11CzSN1LE"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PQczgWXEFtuW",
        "outputId": "4cd6a15d-37ee-4b1e-cc9b-603aa045384f"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.22 (BiocManager 1.30.26), R 4.5.2 (2025-10-31)\n",
            "\n",
            "Installing package(s) 'BiocVersion', 'DESeq2'\n",
            "\n",
            "also installing the dependencies ‘XVector’, ‘formatR’, ‘abind’, ‘SparseArray’, ‘lambda.r’, ‘futile.options’, ‘Seqinfo’, ‘S4Arrays’, ‘DelayedArray’, ‘futile.logger’, ‘snow’, ‘BH’, ‘S4Vectors’, ‘IRanges’, ‘GenomicRanges’, ‘SummarizedExperiment’, ‘BiocGenerics’, ‘Biobase’, ‘BiocParallel’, ‘matrixStats’, ‘locfit’, ‘MatrixGenerics’, ‘RcppArmadillo’\n",
            "\n",
            "\n"
          ]
        }
      ],
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "if (!require(\"DESeq2\", quietly = TRUE))\n",
        "    BiocManager::install(\"DESeq2\")"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "if (!requireNamespace(\"remotes\", quietly = TRUE))\n",
        "    install.packages(\"remotes\")\n",
        "if (!requireNamespace(\"EnhancedVolcano\", quietly = TRUE))\n",
        "    remotes::install_github(\"kevinblighe/EnhancedVolcano\")"
      ],
      "metadata": {
        "id": "cxQWhZrlHEog"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Load R packages"
      ],
      "metadata": {
        "id": "IMnE7X84N9yZ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "library(DESeq2)\n",
        "library(EnhancedVolcano)"
      ],
      "metadata": {
        "id": "Sj36uiwsF_eZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Run DESeq2 to identify DEG"
      ],
      "metadata": {
        "id": "j74SgNjMAUtd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "dds <- DESeqDataSetFromMatrix(countData = merged_df,\n",
        "                              colData = coldata_df,\n",
        "                              design =~ group)"
      ],
      "metadata": {
        "id": "t91b0RvtLK6W"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "class(dds)"
      ],
      "metadata": {
        "id": "RtJKn1kmTpLT"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "The `DESeq()` function normalizes the read counts,estimates dispersions, and fits the linear model, all in one go."
      ],
      "metadata": {
        "id": "1U-O5OY8jxhf"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "dds <- DESeq(dds)"
      ],
      "metadata": {
        "id": "7U_yyOBKT439"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Dispersion is a measure of spread or variability in the data. Variance, standard deviation, IQR, among other measures, can all be used to measure dispersion.\n",
        "\n",
        "DESeq2 uses a specific measure of dispersion (α) related to the mean (μ) and variance of the data: Var = μ + α*μ^2.\n",
        "\n"
      ],
      "metadata": {
        "id": "6ab3VkWcebnP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "sizeFactors(dds)"
      ],
      "metadata": {
        "id": "_ajaOW1J3qRS"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "head(counts(dds,  normalized = TRUE))"
      ],
      "metadata": {
        "id": "1kLsIacY5-wK"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "The function `plotDispEsts` shows the dispersion by mean of normalized counts. We expect the dispersion to decrease as the mean of normalized counts increases.\n",
        "\n",
        "The functions shows:\n",
        "\n",
        "1. black per-gene dispersion estimates\n",
        "\n",
        "2. a red trend line representing the global relationship between dispersion and normalized count\n",
        "\n",
        "3. blue 'shrunken' values moderating individual dispersion estimates by the global relationship\n",
        "\n",
        "4. blue-circled dispersion outliers with high gene-wise dispersion that were not adjusted.\n"
      ],
      "metadata": {
        "id": "uXivh8v3T5GY"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "plotDispEsts(dds)"
      ],
      "metadata": {
        "id": "SwpDLVrvT2uj"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "jWaDenf8XS6x"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Plot normalized genes\n",
        "\n",
        "The function `plotCounts` is used to plot normalized counts plus a pseudocount of 0.5 by default."
      ],
      "metadata": {
        "id": "uQF7TbwWXTVV"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "vsd <- vst(dds, blind=FALSE)"
      ],
      "metadata": {
        "id": "QZUkERccMHUq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "class(vsd)"
      ],
      "metadata": {
        "id": "B57aLAb_WbfO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "plotPCA(vsd, intgroup = c(\"group\"))"
      ],
      "metadata": {
        "id": "zzBpH-CtMaZG"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Extract DEG results using `results` function\n"
      ],
      "metadata": {
        "id": "VPDytRwwQlGM"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "res <- results(dds)\n",
        "res"
      ],
      "metadata": {
        "id": "cD4NTIzUQloD"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "class(res)"
      ],
      "metadata": {
        "id": "h7xDejiVYtod"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "mcols(res)$description"
      ],
      "metadata": {
        "id": "FPiQ4zFFfAUx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "* `baseMean`: mean of normalized counts for all samples\n",
        "* `log2FoldChange`: log2 fold change\n",
        "* `lfcSE`: standard error\n",
        "* `stat`: Wald statistic\n",
        "* `pvalue`: Wald test p-value\n",
        "* `padj`: BH adjusted p-values"
      ],
      "metadata": {
        "id": "hsuhRybGRMJ5"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "If we used the p-value directly from the Wald test with a significance cut-off of p < 0.05, that means there is a 5% chance it is a false positives. Each p-value is the result of a single test (single gene). The more genes we test, the more we inflate the false positive rate. This is the multiple testing problem. For example, if we test 20,000 genes for differential expression, at p < 0.05 we would expect to find 1,000 genes by chance. If we found 3000 genes to be differentially expressed total, roughly one third of our genes are false positives. We would not want to sift through our “significant” genes to identify which ones are true positives.\n",
        "\n",
        "DESeq2 helps reduce the number of genes tested by removing those genes unlikely to be significantly DE prior to testing, such as those with low number of counts and outlier samples (gene-level QC). However, we still need to correct for multiple testing to reduce the number of false positives, and there are a few common approaches:\n",
        "\n",
        "1. Bonferroni: The adjusted p-value is calculated by: p-value * m (m = total number of tests). This is a very conservative approach with a high probability of false negatives, so is generally not recommended.\n",
        "\n",
        "2. FDR/Benjamini-Hochberg: Benjamini and Hochberg (1995) defined the concept of FDR and created an algorithm to control the expected FDR below a specified level given a list of independent p-values. An interpretation of the BH method for controlling the FDR is implemented in DESeq2 in which we rank the genes by p-value, then multiply each ranked p-value by m/rank.\n",
        "\n",
        "3. Q-value / Storey method: The minimum FDR that can be attained when calling that feature significant. For example, if gene X has a q-value of 0.013 it means that 1.3% of genes that show p-values at least as small as gene X are false positives\n",
        "In DESeq2, the p-values attained by the Wald test are corrected for multiple testing using the Benjamini and Hochberg method by default. There are options to use other methods in the results() function. The p-adjusted values should be used to determine significant genes. The significant genes can be output for visualization and/or functional analysis.\n",
        "\n",
        "So what does FDR < 0.05 mean? By setting the FDR cutoff to < 0.05, we’re saying that the proportion of false positives we expect amongst our differentially expressed genes is 5%. For example, if you call 500 genes as differentially expressed with an FDR cutoff of 0.05, you expect 25 of them to be false positives."
      ],
      "metadata": {
        "id": "3fYci13sRU3q"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Note on p-values set to `NA`: some values in the results table can be set to `NA` for one of the following reasons:\n",
        "\n",
        "1. If within a row, all samples have zero counts, the baseMean column will be zero, and the log2 fold change estimates, p value and adjusted p value will all be set to NA.\n",
        "\n",
        "2. If a row contains a sample with an extreme count outlier then the p value and adjusted p value will be set to NA. These outlier counts are detected by Cook’s distance. Customization of this outlier filtering and description of functionality for replacement of outlier counts and refitting is described below\n",
        "\n",
        "3. If a row is filtered by automatic independent filtering, for having a low mean normalized count, then only the adjusted p value will be set to NA. Description and customization of independent filtering is described below"
      ],
      "metadata": {
        "id": "qPh1oDo-fSAk"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "plotMA(res, ylim=c(-2,2))\n"
      ],
      "metadata": {
        "id": "xGoKFbAOYnAs"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "write.csv(res, file = \"BIOI_bulkRNAseq_SE_DESeq2_res.csv\")"
      ],
      "metadata": {
        "id": "28HAyS61YzQC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "d <- plotCounts(dds, gene=which.min(res$padj), intgroup=\"group\",\n",
        "                returnData=TRUE)"
      ],
      "metadata": {
        "id": "hQrotjbAY6Do"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "d"
      ],
      "metadata": {
        "id": "xNIm9Wvla1vr"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "which.min(res$padj)"
      ],
      "metadata": {
        "id": "OhvuXOoSapj6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "library(\"ggplot2\")\n",
        "ggplot(d, aes(x=group, y=count)) +\n",
        "  geom_point(position=position_jitter(w=0.1,h=0)) +\n",
        "  scale_y_log10(breaks=c(25,100,400))"
      ],
      "metadata": {
        "id": "xM5eUKaxY9VF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "  EnhancedVolcano(res,\n",
        "    lab = rownames(res),\n",
        "    x = 'log2FoldChange',\n",
        "    y = 'pvalue')"
      ],
      "metadata": {
        "id": "J6fF6HfLY_Ua"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Understand normalized count in DESeq2 (Optional)"
      ],
      "metadata": {
        "id": "jZKIKSIG2O1q"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "####  Create a pseudo-reference sample (row-wise geometric mean)"
      ],
      "metadata": {
        "id": "gkojMLPP5OTA"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        " The code below creates a new column called `pseudo_reference` that contains the average log-transformed expression value for each gene across all samples. This pseudo-reference is similar to calculating a \"reference sample\" to compare other samples."
      ],
      "metadata": {
        "id": "4zLA4J3IP_XR"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "log_data = log(merged_df)\n",
        "head(log_data)"
      ],
      "metadata": {
        "id": "h5mtnXXgPbxN"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "library(dplyr)\n",
        "library(tibble) # rownames_to_column\n",
        "\n",
        "log_data = log_data %>%\n",
        "             rownames_to_column('gene') %>%\n",
        "             mutate (pseudo_reference = rowMeans(log_data))\n",
        "\n",
        "head(log_data)"
      ],
      "metadata": {
        "id": "4KSHRDDmPwSt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "table(log_data$pseudo_reference == \"-Inf\")"
      ],
      "metadata": {
        "id": "UaTTBRuoWM-v"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "filtered_log_data = log_data %>% filter(pseudo_reference != \"-Inf\")\n",
        "head(filtered_log_data)"
      ],
      "metadata": {
        "id": "b8K7-wp9Uw1i"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "filtered_log_data$pseudo_reference = exp(filtered_log_data$pseudo_reference)"
      ],
      "metadata": {
        "id": "rK3R78H7bBgM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "head(filtered_log_data)"
      ],
      "metadata": {
        "id": "yghVcB4HNx3i"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "dim(log_data)\n",
        "dim(filtered_log_data)"
      ],
      "metadata": {
        "id": "F4t9hEfNU0Xy"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "8IRDUvFn5kyH"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Calculate ratio between each sample and the pseudo-reference for each gene"
      ],
      "metadata": {
        "id": "vyunx-17AoQb"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        " This step calculates the fold change between each sample and the pseudo-reference for each gene.\n"
      ],
      "metadata": {
        "id": "6WEKx0rl-RCS"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "ratio_data = sweep(exp(filtered_log_data[,2:7]), 1, filtered_log_data[,8], \"/\")\n",
        "head(ratio_data)"
      ],
      "metadata": {
        "id": "32ZWp2x92j5f"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "sWzMFJSZd2Jk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Calculate  scaling factor\n",
        "\n",
        "The code below computes the median fold change for each sample across all genes."
      ],
      "metadata": {
        "id": "35iX4Cocd03W"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "Pl7ADLs6V7cO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "scaling_factors = apply(ratio_data, 2, median, na.rm = TRUE)\n",
        "scaling_factors"
      ],
      "metadata": {
        "id": "Q2bD8fLniTd7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "The `2` indicates that the function is applied to columns, i.e., for each sample."
      ],
      "metadata": {
        "id": "LsGCarcI_ipu"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Normalize the counts\n",
        "\n",
        "This step below normalizes each sample by its scaling factors, making the data comparable across samples. The result is a normalized gene expression matrix."
      ],
      "metadata": {
        "id": "D1qZE5TDBFwQ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "manually_normalized = sweep(merged_df, 2, scaling_factors, \"/\")\n",
        "head(manually_normalized)"
      ],
      "metadata": {
        "id": "qnYw-5Tj3E_L"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "hist(manually_normalized$N2_day1_rep1)"
      ],
      "metadata": {
        "id": "9kmTgcCxgRru"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "The code below shows that the size factors and the normalized read counts calculated by ourselves are the same as what DESeq2 function returns."
      ],
      "metadata": {
        "id": "bvqGGjlLeKRq"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "head(counts(dds,  normalized = TRUE))\n"
      ],
      "metadata": {
        "id": "XpQem8Q7XmA9"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "sizeFactors(dds)"
      ],
      "metadata": {
        "id": "Ux69bagpXnrW"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## SessionInfo"
      ],
      "metadata": {
        "id": "0TChcITuyWWX"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "sessionInfo()"
      ],
      "metadata": {
        "id": "n5Wwjb64yUY8"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}